<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>18&nbsp; Advanced regression topics – Introduction to data analysis and visualization with R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./43_Anova.html" rel="next">
<link href="./41_Linear_regression.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-8da5b4427184b79ecddefad3d342027e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>

<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./40_Intro_adv_statistics.html">Part III: Advanced statistics</a></li><li class="breadcrumb-item"><a href="./42_Multiple_and_Nonlinear_regression.html"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to data analysis and visualization with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Part I: Processing and visualizing data with R</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_Intro_R_RStudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_R_programming_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R programming basics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_Functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_Data_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Reading tabular data from disk</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_Basic_data_wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Basic data manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_Summaries_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Summary statistics and tidy data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_Creating_figures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Creating publication-grade figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_Further_plotting_options.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Some further plotting options; introducing factors</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_Joining_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Joining data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Part II: Introductory statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_Probabilities_and_distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Probabilities and probability distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./22_Sampling_distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Sampling distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_Statistical_inference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to statistical inference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./24_Statistical_inference_in_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Statistical inference in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./25_Statistical_inference_two.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Statistical inference in two groups</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./26_Statistical_inference_three.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Statistical inference in qualitative variables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Part III: Advanced statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_Intro_adv_statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Relationships between variables</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./41_Linear_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Simple linear regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./42_Multiple_and_Nonlinear_regression.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./43_Anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./44_Anova_two_way.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Two-way ANOVA and the Scheirer–Ray–Hare test</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./45_Intro_map.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Higher-order functions and mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./46_Multiple_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Nested data and multiple analysis</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">References</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./99_References.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">References</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-multiple-regression" id="toc-sec-multiple-regression" class="nav-link active" data-scroll-target="#sec-multiple-regression"><span class="header-section-number">18.1</span> Multiple linear regression</a>
  <ul>
  <li><a href="#visualizing-the-response-variable" id="toc-visualizing-the-response-variable" class="nav-link" data-scroll-target="#visualizing-the-response-variable"><span class="header-section-number">18.1.1</span> Visualizing the response variable</a></li>
  <li><a href="#sec-pairwise-visualization" id="toc-sec-pairwise-visualization" class="nav-link" data-scroll-target="#sec-pairwise-visualization"><span class="header-section-number">18.1.2</span> Pairwise Relationships</a>
  <ul class="collapse">
  <li><a href="#quantitative-explanatory-variables" id="toc-quantitative-explanatory-variables" class="nav-link" data-scroll-target="#quantitative-explanatory-variables"><span class="header-section-number">18.1.2.1</span> Quantitative explanatory variables</a></li>
  <li><a href="#other-quantitative-variables" id="toc-other-quantitative-variables" class="nav-link" data-scroll-target="#other-quantitative-variables"><span class="header-section-number">18.1.2.2</span> Other quantitative variables</a></li>
  <li><a href="#qualitative-explanatory-variables" id="toc-qualitative-explanatory-variables" class="nav-link" data-scroll-target="#qualitative-explanatory-variables"><span class="header-section-number">18.1.2.3</span> Qualitative explanatory variables</a></li>
  </ul></li>
  <li><a href="#multiple-linear-regression-model" id="toc-multiple-linear-regression-model" class="nav-link" data-scroll-target="#multiple-linear-regression-model"><span class="header-section-number">18.1.3</span> Multiple linear regression model</a></li>
  <li><a href="#sec-indicator-variable" id="toc-sec-indicator-variable" class="nav-link" data-scroll-target="#sec-indicator-variable"><span class="header-section-number">18.1.4</span> Indicator variables</a></li>
  <li><a href="#model-estimation" id="toc-model-estimation" class="nav-link" data-scroll-target="#model-estimation"><span class="header-section-number">18.1.5</span> Model estimation</a></li>
  <li><a href="#model-evaluation" id="toc-model-evaluation" class="nav-link" data-scroll-target="#model-evaluation"><span class="header-section-number">18.1.6</span> Model evaluation</a></li>
  <li><a href="#model-inference" id="toc-model-inference" class="nav-link" data-scroll-target="#model-inference"><span class="header-section-number">18.1.7</span> Model inference</a>
  <ul class="collapse">
  <li><a href="#sec-seq-ss" id="toc-sec-seq-ss" class="nav-link" data-scroll-target="#sec-seq-ss"><span class="header-section-number">18.1.7.1</span> Sequential Sums of Squares</a></li>
  <li><a href="#partial-f-test-for-groups-of-parameters" id="toc-partial-f-test-for-groups-of-parameters" class="nav-link" data-scroll-target="#partial-f-test-for-groups-of-parameters"><span class="header-section-number">18.1.7.2</span> Partial F-test for groups of parameters</a></li>
  </ul></li>
  <li><a href="#model-metrics" id="toc-model-metrics" class="nav-link" data-scroll-target="#model-metrics"><span class="header-section-number">18.1.8</span> Model metrics</a></li>
  </ul></li>
  <li><a href="#sec-theilsen" id="toc-sec-theilsen" class="nav-link" data-scroll-target="#sec-theilsen"><span class="header-section-number">18.2</span> A non-parametric method: Theil–Sen regression</a></li>
  <li><a href="#sec-nonlin-regression" id="toc-sec-nonlin-regression" class="nav-link" data-scroll-target="#sec-nonlin-regression"><span class="header-section-number">18.3</span> Nonlinear regression</a>
  <ul>
  <li><a href="#sec-lists" id="toc-sec-lists" class="nav-link" data-scroll-target="#sec-lists"><span class="header-section-number">18.3.1</span> Interlude: lists</a></li>
  <li><a href="#back-to-nonlinear-regression" id="toc-back-to-nonlinear-regression" class="nav-link" data-scroll-target="#back-to-nonlinear-regression"><span class="header-section-number">18.3.2</span> Back to nonlinear regression</a></li>
  </ul></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">18.4</span> Exercises</a>
  <ul>
  <li><a href="#multiple-linear-regression" id="toc-multiple-linear-regression" class="nav-link" data-scroll-target="#multiple-linear-regression"><span class="header-section-number">18.4.1</span> Multiple linear regression</a></li>
  <li><a href="#non-parametric-regression" id="toc-non-parametric-regression" class="nav-link" data-scroll-target="#non-parametric-regression"><span class="header-section-number">18.4.2</span> Non-parametric regression</a></li>
  <li><a href="#sec-exercises-non-linear" id="toc-sec-exercises-non-linear" class="nav-link" data-scroll-target="#sec-exercises-non-linear"><span class="header-section-number">18.4.3</span> Non-linear regression</a>
  <ul class="collapse">
  <li><a href="#exponential-growth" id="toc-exponential-growth" class="nav-link" data-scroll-target="#exponential-growth"><span class="header-section-number">18.4.3.1</span> Exponential growth</a></li>
  <li><a href="#nitrogen-uptake" id="toc-nitrogen-uptake" class="nav-link" data-scroll-target="#nitrogen-uptake"><span class="header-section-number">18.4.3.2</span> Nitrogen uptake</a></li>
  <li><a href="#logistic-growth" id="toc-logistic-growth" class="nav-link" data-scroll-target="#logistic-growth"><span class="header-section-number">18.4.3.3</span> Logistic growth</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./40_Intro_adv_statistics.html">Part III: Advanced statistics</a></li><li class="breadcrumb-item"><a href="./42_Multiple_and_Nonlinear_regression.html"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span id="sec-adv-reg" class="quarto-section-identifier"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In the vast majority of cases, a simple linear regression model will not be sufficient to model a phenomena of a response variable. Instead the more general case involves multiple explanatory variables of different types that together allow us to describe the response. We can extend the theory presented in <a href="41_Linear_regression.html" class="quarto-xref"><span>Chapter 17</span></a> to <strong>multiple linear regression</strong>, expand to <strong>non-linear regression</strong> and <strong>non-parametric regression</strong> methods that does not require the same assumptions of the data or model.</p>
<section id="sec-multiple-regression" class="level2" data-number="18.1">
<h2 data-number="18.1" class="anchored" data-anchor-id="sec-multiple-regression"><span class="header-section-number">18.1</span> Multiple linear regression</h2>
<p>Following the steps in simple linear regression we begin with visualizing the data. In the case of multiple linear regression we now have a set of possible explanatory variables at our disposal to visualize and assess their viability in the model.</p>
<p>As an example we will be using the <code>penguins</code> dataset collected by a research team in Antarctica between 2007 and 2009. The team gathered information on 333 penguins across three islands near the Palmer Research Station. The dataset is available via the <code>palmerpenguins</code> package <span class="citation" data-cites="penguins2020">(<a href="99_References.html#ref-penguins2020" role="doc-biblioref">Horst, Hill, and Gorman 2020</a>)</span> and can be loaded into R using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Don’t forget to install the package if you haven’t already</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("palmerpenguins")</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the package with the dataset</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(palmerpenguins)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter out observations with missing values</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>penguins <span class="ot">&lt;-</span> </span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  penguins <span class="sc">|&gt;</span> </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(sex))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>As part of the data exploration, we can identify missing values in some variables and we choose to filter them out in this example. Handling missing values is a large field in statistics, and there are several methods that can <em>impute</em> (estimate the missing value) so we don’t have to remove entire observations from the analysis.</p>
<p>A simple imputation method is mean imputation, where we replace the missing value with the mean of the other measurements, or the mode if a qualitative variable is being imputed. In practice, more advanced methods are used that can be applied in many different cases and also take other information about the observations into account.</p>
</div>
</div>
<p>We can take a closer look at a sample of the dataset in <a href="#tbl-penguins-sample" class="quarto-xref">Table&nbsp;<span>18.1</span></a>.</p>
<div class="cell" data-cap-location="top">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a formatted table using kable() and style it with kable_styling()</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">5</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>() <span class="sc">|&gt;</span> </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-penguins-sample" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-penguins-sample-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;18.1: Sample of observations from the dataset.
</figcaption>
<div aria-describedby="tbl-penguins-sample-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> species </th>
   <th style="text-align:left;"> island </th>
   <th style="text-align:right;"> bill_length_mm </th>
   <th style="text-align:right;"> bill_depth_mm </th>
   <th style="text-align:right;"> flipper_length_mm </th>
   <th style="text-align:right;"> body_mass_g </th>
   <th style="text-align:left;"> sex </th>
   <th style="text-align:right;"> year </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> Adelie </td>
   <td style="text-align:left;"> Torgersen </td>
   <td style="text-align:right;"> 39.1 </td>
   <td style="text-align:right;"> 18.7 </td>
   <td style="text-align:right;"> 181 </td>
   <td style="text-align:right;"> 3750 </td>
   <td style="text-align:left;"> male </td>
   <td style="text-align:right;"> 2007 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Adelie </td>
   <td style="text-align:left;"> Torgersen </td>
   <td style="text-align:right;"> 39.5 </td>
   <td style="text-align:right;"> 17.4 </td>
   <td style="text-align:right;"> 186 </td>
   <td style="text-align:right;"> 3800 </td>
   <td style="text-align:left;"> female </td>
   <td style="text-align:right;"> 2007 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Adelie </td>
   <td style="text-align:left;"> Torgersen </td>
   <td style="text-align:right;"> 40.3 </td>
   <td style="text-align:right;"> 18.0 </td>
   <td style="text-align:right;"> 195 </td>
   <td style="text-align:right;"> 3250 </td>
   <td style="text-align:left;"> female </td>
   <td style="text-align:right;"> 2007 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Adelie </td>
   <td style="text-align:left;"> Torgersen </td>
   <td style="text-align:right;"> 36.7 </td>
   <td style="text-align:right;"> 19.3 </td>
   <td style="text-align:right;"> 193 </td>
   <td style="text-align:right;"> 3450 </td>
   <td style="text-align:left;"> female </td>
   <td style="text-align:right;"> 2007 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Adelie </td>
   <td style="text-align:left;"> Torgersen </td>
   <td style="text-align:right;"> 39.3 </td>
   <td style="text-align:right;"> 20.6 </td>
   <td style="text-align:right;"> 190 </td>
   <td style="text-align:right;"> 3650 </td>
   <td style="text-align:left;"> male </td>
   <td style="text-align:right;"> 2007 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
</div>
<p>From the table, we can identify the following variables:</p>
<ul>
<li><code>species</code>: The penguin’s species is a qualitative variable. We cannot say one species is “better” or “larger” than another. The categories cannot be ranked, so this variable follows a nominal scale.</li>
<li><code>island</code>: The island where the penguin was located at the time of measurement is also a qualitative variable that cannot be ranked. This variable also follows a nominal scale.</li>
<li><code>bill_length_mm</code>: A quantitative variable measuring the bill length in millimeters (mm). Length is a typical variable that follows a ratio scale because it has a clear zero point.</li>
<li><code>bill_depth_mm</code>: Measures the bill depth in millimeters and follows the same reasoning as bill length.</li>
<li><code>flipper_length_mm</code>: Another variable measuring length, this time the flipper length. Same reasoning applies.</li>
<li><code>body_mass_g</code>: The penguin’s weight in grams. Weight also has a clear zero point and is considered quantitative, following a ratio scale.</li>
<li><code>sex</code>: The penguin’s biological sex is a qualitative variable that cannot be ranked—nominal scale.</li>
<li><code>year</code>: This variable is a bit harder to assess. It measures the year the penguin was measured as a numeric variable (integers, so R stores it as <code>int</code>), but in this context, it may not be considered quantitative. For simplicity, we can say that since it’s possible to calculate differences between years (e.g., 1 year between 2007 and 2008), but there’s no clear zero point, this variable can be considered quantitative and follows an interval scale.</li>
</ul>
<p>We will focus on bill length as our response variable in the following examples. Since the dataset is from an observational study, we cannot draw conclusions about causal relationships, but we can investigate correlations between the penguins’ various characteristics.</p>
<section id="visualizing-the-response-variable" class="level3" data-number="18.1.1">
<h3 data-number="18.1.1" class="anchored" data-anchor-id="visualizing-the-response-variable"><span class="header-section-number">18.1.1</span> Visualizing the response variable</h3>
<p>As a first step in the exploratory analysis, we can visualize the distribution of the response variable using a histogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(penguins) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> bill_length_mm) <span class="sc">+</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">30</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Bill Length (mm)"</span>, <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-hist-y" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-hist-y-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-hist-y-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-hist-y-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.1: Histogram of bill length distribution
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-hist-y" class="quarto-xref">Figure&nbsp;<span>18.1</span></a> gives us a picture of the variable’s characteristics and whether the data contains any outliers that might be difficult to capture with a model. Bill length appears to have a bimodal structure with two centers of mass around 38–40 and 50 mm. Most observations fall between approximately 35–52 mm, but there are a few around 58–60 mm that seem unusually large.</p>
<p>The fact that the distribution doesn’t look normal doesn’t matter, we need to examine the distribution of the response variable <em>with respect to</em> the explanatory variables to check a regression model’s assumptions.</p>
</section>
<section id="sec-pairwise-visualization" class="level3" data-number="18.1.2">
<h3 data-number="18.1.2" class="anchored" data-anchor-id="sec-pairwise-visualization"><span class="header-section-number">18.1.2</span> Pairwise Relationships</h3>
<p>The dataset contains several potential explanatory variables that could be included in a model. Depending on how a study was conducted, variables may be excluded if they don’t have a logical relationship with the response variable, for example, ID variables are not relevant to investigate. In our example data, there is a variable describing the year the observation was made, which we can initially assume doesn’t have a logical relationship with bill length. That leaves six other variables that could be included in the model fitting.</p>
<section id="quantitative-explanatory-variables" class="level4" data-number="18.1.2.1">
<h4 data-number="18.1.2.1" class="anchored" data-anchor-id="quantitative-explanatory-variables"><span class="header-section-number">18.1.2.1</span> Quantitative explanatory variables</h4>
<p>Similar to <a href="41_Linear_regression.html#fig-shape-size" class="quarto-xref">Figure&nbsp;<span>17.1</span></a> we create scatter plots between each pair of continuous explanatory variable and the continuous response and investigate the four questions about the relationship.</p>
<ol type="1">
<li>Is the relationship linear?</li>
<li>Is the relationship positive or negative?</li>
<li>Is the relationship strong or weak?</li>
<li>Are there any outliers?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(penguins) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> bill_length_mm) <span class="sc">+</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span> </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Body Weight (g)"</span>, <span class="at">y =</span> <span class="st">"Bill Length (mm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-weight-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-weight-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-weight-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-weight-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.2: Scatterplot showing the relationship between body weight and bill length
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-weight-length" class="quarto-xref">Figure&nbsp;<span>18.2</span></a> shows that the relationship appears mostly linear, as a constant change (increase) in body weight leads to a constant change (increase) in bill length. Most points seem to follow this trend, indicating a relatively strong relationship. However, there are several observations (highlighted in <a href="#fig-w-l-ex" class="quarto-xref">Figure&nbsp;<span>18.3</span></a>) that deviate from this. These observations have lower body weight but the same bill length as penguins with higher body weight, which affects the strength of the relationship.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-w-l-ex" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-w-l-ex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-w-l-ex-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-w-l-ex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.3: Scatterplot with highlighted area in the circle
</figcaption>
</figure>
</div>
</div>
</div>
<p><span class="math display">\[
r = 0.589
\]</span></p>
<p>Since the correlation coefficient is close to 0.6, it suggests the relationship is moderately strong.</p>
<p><a href="#fig-weight-length" class="quarto-xref">Figure&nbsp;<span>18.2</span></a> also shows some observations that could be considered outliers. For example, <span class="math inline">\(\{x = ~2700, y = ~47\}\)</span> and <span class="math inline">\(\{x = ~3700, y = ~58\}\)</span> are observations that deviate significantly from the expected relationship and other observations. A detailed analysis of outliers will be covered in later chapters, but for exploratory purposes, we note that some observations may affect model fitting.</p>
<p>In summary, the relationship between body weight and bill length is:</p>
<ol type="1">
<li>linear,<br>
</li>
<li>positive,<br>
</li>
<li>moderately strong,<br>
</li>
<li>with possible outliers.</li>
</ol>
<p>When we build our first model, it will likely be sufficient to include a simple <span class="math inline">\(\beta_1 \cdot \text{body weight}\)</span> term.</p>
</section>
<section id="other-quantitative-variables" class="level4" data-number="18.1.2.2">
<h4 data-number="18.1.2.2" class="anchored" data-anchor-id="other-quantitative-variables"><span class="header-section-number">18.1.2.2</span> Other quantitative variables</h4>
<p>The same exploration should be done for all quantitative variable pairs, in this case also bill depth and flipper length:</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-width-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-width-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-width-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-width-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.4: Scatterplot showing the relationship between bill depth and bill length
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-flipper-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-flipper-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-flipper-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-flipper-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.5: Scatterplot showing the relationship between flipper length and bill length
</figcaption>
</figure>
</div>
</div>
</div>
<p>Both variables appear to have a linear relationship with the response variable. <a href="#fig-width-length" class="quarto-xref">Figure&nbsp;<span>18.4</span></a> suggests that the relationship between bill depth and bill length is weakly negative (<span class="math inline">\(r = -0.229\)</span>) as the points are widely scattered, while <a href="#fig-flipper-length" class="quarto-xref">Figure&nbsp;<span>18.5</span></a> suggests a somewhat stronger positive relationship (<span class="math inline">\(r = 0.653\)</span>), similar to <a href="#fig-weight-length" class="quarto-xref">Figure&nbsp;<span>18.2</span></a>.</p>
<p>A new phenomenon visible in <a href="#fig-width-length" class="quarto-xref">Figure&nbsp;<span>18.4</span></a> is that we seem to have several clusters of points, each with a positive relationship, even though the overall relationship was interpreted as weakly negative. If we had only calculated the correlation coefficient, this phenomenon would have escaped our analysis. <a href="#fig-width-length-annot" class="quarto-xref">Figure&nbsp;<span>18.6</span></a> is an example of <em>Simpson’s Paradox</em>, which we will explore further later in this material.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-width-length-annot" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-width-length-annot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-width-length-annot-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-width-length-annot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.6: Groupings of observations
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="qualitative-explanatory-variables" class="level4" data-number="18.1.2.3">
<h4 data-number="18.1.2.3" class="anchored" data-anchor-id="qualitative-explanatory-variables"><span class="header-section-number">18.1.2.3</span> Qualitative explanatory variables</h4>
<p>We cannot use scatter plots to visualize the relationship between qualitative explanatory variables and a continuous response variable. Instead, we need visualizations that account for the qualitative scale, usually ordinal or nominal. There are several ways to visualize the distribution of the response variable across levels of the explanatory variable, such as grouped histograms or box plots, but one type of visualization that shows distribution details is a <em>violin plot</em>. A violin plot is a mirrored density plot, where areas with many observations have a larger area under the curve.</p>
<p>Using <code>ggplot2</code>, we can create such a plot with <code>geom_violin()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(penguins) <span class="sc">+</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> bill_length_mm) <span class="sc">+</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Species"</span>, <span class="at">y =</span> <span class="st">"Bill Length (mm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-species-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-species-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-species-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-species-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.7: Distribution of bill length by species
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-species-length" class="quarto-xref">Figure&nbsp;<span>18.7</span></a> shows that Adelie penguins generally have shorter bill lengths compared to Chinstrap and Gentoo, as the center of the distribution is around 38–40 mm. Chinstrap penguins have a slightly larger proportion with bill lengths over 50 mm, while Gentoo have a larger proportion under 50 mm.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-sex-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-sex-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-sex-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-sex-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.8: Distribution of bill length by sex
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-sex-length" class="quarto-xref">Figure&nbsp;<span>18.8</span></a> has a slightly different shape, with two large masses for each category. This likely indicates that sex within species has an effect, and that males generally have longer bills than females of the same species.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-island-length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-island-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-island-length-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-island-length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.9: Distribution of bill length by island
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-island-length" class="quarto-xref">Figure&nbsp;<span>18.9</span></a> suggests that penguins on Torgersen Island have shorter bill lengths than those on other islands. However, we need to consider whether this variable actually describes the relationship or if another phenomenon explains it, for example if one species is more prevalent on one specific island, like a confounder/control.</p>
<p>The conclusion from these visualizations is that there appears to be a relationship between species and sex with bill length, and these two variables should be included in the model. We now need to consider how to appropriately include a qualitative variable containing text in a mathematical model that requires numerical input.</p>
</section>
</section>
<section id="multiple-linear-regression-model" class="level3" data-number="18.1.3">
<h3 data-number="18.1.3" class="anchored" data-anchor-id="multiple-linear-regression-model"><span class="header-section-number">18.1.3</span> Multiple linear regression model</h3>
<p>After summarizing observations from visualizations and descriptive statistics, the next step in the process is to build the structure of the model. The simplest approach is to start with the quantitative variables, which (usually) only require one term each in the model. <span id="eq-penguin-quant"><span class="math display">\[
\begin{aligned}
  \text{Bill Length} = &amp;\beta_0 + \beta_1 \cdot \text{Body Weight} + \beta_2 \cdot \text{Flipper Length} + \\
&amp;\beta_3 \cdot \text{Bill Depth} + \cdots + \varepsilon
\end{aligned}
\tag{18.1}\]</span></span></p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Regardless of whether the model includes one or several explanatory variables, we must always keep in mind the five assumptions presented in <a href="40_Intro_adv_statistics.html#sec-model-assumptions" class="quarto-xref"><span>Section 16.1.1</span></a>, especially the assumption of linearity. If we’ve discovered non-linear relationships in the pairwise visualizations, it’s often not enough to include just a single term in the model, but we might need to look at transformations.</p>
</div>
</div>
</section>
<section id="sec-indicator-variable" class="level3" data-number="18.1.4">
<h3 data-number="18.1.4" class="anchored" data-anchor-id="sec-indicator-variable"><span class="header-section-number">18.1.4</span> Indicator variables</h3>
<p>A regression model cannot directly handle qualitative variables, such as <span class="math inline">\(\beta_4 \cdot \text{Species}\)</span>, because the variable’s values represent categories, not values from a numerical scale. This applies even if the qualitative variable is numerically coded, e.g.&nbsp;1, 2, 3 for each species. A slope parameter describes the constant change in the response variable when the corresponding explanatory variable increases by one unit, but a qualitative variable usually has no unit and no constant changes between adjacent values. Instead, we must transform the qualitative variable into numeric form using <em>indicator variables</em> (also called dummy variables).</p>
<p>As the name suggests, <strong>indicator</strong> variables are used to <strong>indicate</strong> which category an observation has for the qualitative variable. We need to create a limited number of indicator variables that clearly show exactly one category per observation.</p>
<p>Assume a qualitative variable has 3 categories: <span class="math display">\[
\begin{bmatrix}A\\B\\C\end{bmatrix}
\]</span> We can start by creating an indicator variable for category A that takes the value 1 if the observation has category A, and 0 otherwise:</p>
<p><span class="math display">\[
\begin{bmatrix}A\\B\\C\end{bmatrix} = \begin{bmatrix}1\\0\\0\end{bmatrix}
\]</span> With only one indicator variable, we cannot clearly identify whether an observation has category B or C, since both have the value 0. So we add another indicator variable that takes the value 1 if the observation has category B, and 0 otherwise:</p>
<p><span class="math display">\[
\begin{bmatrix}A\\B\\C\end{bmatrix} = \begin{bmatrix}1 &amp; 0\\0 &amp; 1\\0 &amp; 0\end{bmatrix}
\]</span> It might seem natural to create an indicator for the last category as well, but it’s not necessary. If both indicators are 0, we’ve successfully identified that the observation has category C, and another variable would just be redundant.</p>
<p>The last category also becomes our <em>reference category</em>, the category against which the effects of the other indicator variables are interpreted. When interpreting slope parameters for indicator variables — e.g., the indicator for A — the change in <span class="math inline">\(Y\)</span> is measured when <span class="math inline">\(X = A\)</span> compared to when <span class="math inline">\(X = C\)</span>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Mathematically, creating three indicator variables would model a perfect relationship and cause singularity issues in the calculations.</p>
</div>
</div>
<p>In general, we create <span class="math inline">\(\text{number of categories} - 1\)</span> indicator variables for each qualitative variable to be included in a regression model. The choice of reference category is arbitrary, but typically the first or last category is used.</p>
<p>To complete the modeling from <a href="#eq-penguin-quant" class="quarto-xref">Equation&nbsp;<span>18.1</span></a>, we’ll include Species and Sex in the model. We need to create two sets of indicator variables with two and one indicators respectively:</p>
<p><span class="math display">\[\begin{align*}
  Gentoo &amp;= \begin{cases}
            1 \quad \text{if species is Gentoo}\\
            0 \quad \text{otherwise}
        \end{cases}\\
  Chinstrap &amp;= \begin{cases}
      1 \quad \text{if species is Chinstrap}\\
      0 \quad \text{otherwise}
  \end{cases}
\end{align*}\]</span></p>
<p>and</p>
<p><span class="math display">\[\begin{align*}
  Male &amp;= \begin{cases}
            1 \quad \text{if male}\\
            0 \quad \text{otherwise}
        \end{cases}
\end{align*}\]</span></p>
<p>to finally create the following model: <span id="eq-penguin-full-simple"><span class="math display">\[
\begin{aligned}
  \text{Bill Length} = &amp;\beta_0 + \beta_1 \cdot \text{Body Weight} + \beta_2 \cdot \text{Flipper Length} + \\
  &amp;\beta_3 \cdot \text{Bill Depth} + \beta_4 \cdot \text{Gentoo} + \beta_5 \cdot \text{Chinstrap} + \\
  &amp;\beta_6 \cdot \text{Male} + \varepsilon
\end{aligned}
\tag{18.2}\]</span></span> where Adelie and Female act as the reference categories for each qualitative variable.</p>
</section>
<section id="model-estimation" class="level3" data-number="18.1.5">
<h3 data-number="18.1.5" class="anchored" data-anchor-id="model-estimation"><span class="header-section-number">18.1.5</span> Model estimation</h3>
<p>We can denote the estimated model with its regression coefficients as:</p>
<p><span id="eq-penguin-est-simple"><span class="math display">\[
\begin{aligned}
  \hat{y}_i = &amp;b_0 + b_1 \cdot x_{1i} + b_2 \cdot x_{2i} + b_3 \cdot x_{3i} + \\
  &amp;b_4 \cdot x_{4i} + b_5 \cdot x_{5i} + b_6 \cdot x_{6i}
\end{aligned}
\tag{18.3}\]</span></span></p>
<p>where <span class="math display">\[\begin{align*}
  \hat{y}_i &amp;= \text{estimated value of the response variable for observation } i\\
  b_0 &amp;= \text{estimate of the intercept}\\
  b_1 - b_6 &amp;= \text{estimates of the slope parameters}
\end{align*}\]</span></p>
<p>To estimate this linear regression model in R, we again make use of the <code>lm()</code> function but adjust the formula argument to encompass more than one variable. Generally, the format is <code>y ~ x</code>, where <code>x</code> consists of the explanatory variables, for example <code>bill_length_mm ~ body_mass_g + bill_depth_mm</code> but there exist a shorthand (<code>~ .</code>) that makes use of all other variables present in the data in the right hand side.</p>
<p>In order for this to work we need to select only the variables we want to include in the model before AND we must also ensure that all variables in the dataset have the correct variable type as expected. We identified in <a href="41_Linear_regression.html#sec-example-data" class="quarto-xref"><span>Section 17.1</span></a> that we had three quantitative variables and two qualitative variables, which in R correspond to the types <code>numeric</code> and <code>Factor</code>. Using <code>Factor</code> simplifies the transformation into indicator variables because R knows it must do so for the model to work. If the qualitative variables were of type <code>character</code> or coded as <code>numeric</code>, R might not create indicator variables and the estimated model would be misinterpreted. We can inspect the variable types in <code>penguins</code> using <code>str()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Only include variables that were found to be related to the response variable</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>modelData <span class="ot">&lt;-</span> </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  penguins <span class="sc">|&gt;</span> </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    species,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    bill_depth_mm,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    flipper_length_mm,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    body_mass_g,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    sex,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    bill_length_mm</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the specified model</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>simpleModel <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> bill_length_mm <span class="sc">~</span> ., <span class="at">data =</span> modelData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tbl-cap-location="top">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(simpleModel) <span class="sc">|&gt;</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coef</span>() <span class="sc">|&gt;</span> </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="cn">NA</span>) <span class="sc">|&gt;</span> </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>() <span class="sc">|&gt;</span> </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at"> </span><span class="st">`</span> <span class="ot">=</span> rowname,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">Estimate =</span> Estimate,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">StdError =</span> <span class="st">`</span><span class="at">Std. Error</span><span class="st">`</span>,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">t-value</span><span class="st">`</span> <span class="ot">=</span> <span class="st">`</span><span class="at">t value</span><span class="st">`</span>,</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">`</span><span class="at">p-value</span><span class="st">`</span> <span class="ot">=</span> <span class="st">`</span><span class="at">Pr(&gt;|t|)</span><span class="st">`</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>(</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">digits =</span> <span class="dv">4</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-reg-summary-good" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-reg-summary-good-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;18.2: A clean display of the model’s estimated parameters
</figcaption>
<div aria-describedby="tbl-reg-summary-good-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">  </th>
   <th style="text-align:right;"> Estimate </th>
   <th style="text-align:right;"> StdError </th>
   <th style="text-align:right;"> t-value </th>
   <th style="text-align:right;"> p-value </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> (Intercept) </td>
   <td style="text-align:right;"> 15.0166 </td>
   <td style="text-align:right;"> 4.3742 </td>
   <td style="text-align:right;"> 3.4330 </td>
   <td style="text-align:right;"> 0.0007 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> speciesChinstrap </td>
   <td style="text-align:right;"> 9.5655 </td>
   <td style="text-align:right;"> 0.3497 </td>
   <td style="text-align:right;"> 27.3508 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> speciesGentoo </td>
   <td style="text-align:right;"> 6.4044 </td>
   <td style="text-align:right;"> 1.0304 </td>
   <td style="text-align:right;"> 6.2154 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> bill_depth_mm </td>
   <td style="text-align:right;"> 0.3130 </td>
   <td style="text-align:right;"> 0.1541 </td>
   <td style="text-align:right;"> 2.0316 </td>
   <td style="text-align:right;"> 0.0430 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> flipper_length_mm </td>
   <td style="text-align:right;"> 0.0686 </td>
   <td style="text-align:right;"> 0.0232 </td>
   <td style="text-align:right;"> 2.9608 </td>
   <td style="text-align:right;"> 0.0033 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> body_mass_g </td>
   <td style="text-align:right;"> 0.0011 </td>
   <td style="text-align:right;"> 0.0004 </td>
   <td style="text-align:right;"> 2.5617 </td>
   <td style="text-align:right;"> 0.0109 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> sexmale </td>
   <td style="text-align:right;"> 2.0297 </td>
   <td style="text-align:right;"> 0.3892 </td>
   <td style="text-align:right;"> 5.2153 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
</div>
<p><a href="#tbl-reg-summary-good" class="quarto-xref">Table&nbsp;<span>18.2</span></a> shows the estimated slope parameters (coefficients). For example, we can see that for each additional gram a penguin weighs, the bill length increases by approximately 0.0011 mm on average <strong>given that all other variables are held constant</strong>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>This last part of the interpretation must be included in a multiple linear regression, as a change in multiple variables would result in a combined change in the response variable relative to each coefficient. In order to isolate one variable’s relationship with the response, we need to isolate its interpretation as well.</p>
</div>
</div>
<p>Indicator variables are interpreted within each set of connected coefficients, relative to the assigned reference category. For example, Gentoo penguins have on average a 6.4 mm longer bill than the reference category Adelie penguins, given that all other variables are held constant.</p>
</section>
<section id="model-evaluation" class="level3" data-number="18.1.6">
<h3 data-number="18.1.6" class="anchored" data-anchor-id="model-evaluation"><span class="header-section-number">18.1.6</span> Model evaluation</h3>
<p>After estimating a model based on observations from visualizations and descriptive statistics, we move on to the residual analysis to determine if th model is appropriate and meets the model assumptions. Using the custom function we designed in <a href="41_Linear_regression.html#sec-diag-plots" class="quarto-xref"><span>Section 17.3.2</span></a> we simply input the estimated model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diagnosticPlots</span>(simpleModel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-penguin-resid" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-penguin-resid-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-penguin-resid-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguin-resid-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.10: Residual analysis for the penguins model
</figcaption>
</figure>
</div>
</div>
</div>
<p>In summary, <a href="#fig-penguin-resid" class="quarto-xref">Figure&nbsp;<span>18.10</span></a> shows that the residuals meet the assumptions of normal distribution with mean 0 and constant variance. There are no clear patterns in any plot that suggest otherwise or that the model fails to capture part of the relationship. A few outliers have been identified, specifically two large positive residuals which might warrant a closer look as to why they are so large, but nothing that deviates from the model assumptions.</p>
</section>
<section id="model-inference" class="level3" data-number="18.1.7">
<h3 data-number="18.1.7" class="anchored" data-anchor-id="model-inference"><span class="header-section-number">18.1.7</span> Model inference</h3>
<p>Using the different sources of variation, we can calculate tests for the whole model or parts of it using various F-tests, while the individual parameter estimates and their associated standard errors can be used in tests for individual slope parameters.</p>
<section id="sec-seq-ss" class="level4" data-number="18.1.7.1">
<h4 data-number="18.1.7.1" class="anchored" data-anchor-id="sec-seq-ss"><span class="header-section-number">18.1.7.1</span> Sequential Sums of Squares</h4>
<p>The calculations for an ANOVA table are performed automatically in R when we use <code>lm()</code>, and we can extract the table from the model object using <code>anova()</code> (see <a href="41_Linear_regression.html#tip-lm-objects" class="quarto-xref">Tip&nbsp;<span>17.1</span></a>).</p>
<div class="cell" data-tbl-cap-location="top">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(simpleModel) <span class="sc">|&gt;</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">4</span>) <span class="sc">|&gt;</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>() <span class="sc">|&gt;</span> </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-anova-example-R" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-anova-example-R-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;18.3: ANOVA table from R
</figcaption>
<div aria-describedby="tbl-anova-example-R-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">  </th>
   <th style="text-align:right;"> Df </th>
   <th style="text-align:right;"> Sum Sq </th>
   <th style="text-align:right;"> Mean Sq </th>
   <th style="text-align:right;"> F value </th>
   <th style="text-align:right;"> Pr(&gt;F) </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> species </td>
   <td style="text-align:right;"> 2 </td>
   <td style="text-align:right;"> 7015.3857 </td>
   <td style="text-align:right;"> 3507.6929 </td>
   <td style="text-align:right;"> 713.4929 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> bill_depth_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 818.5050 </td>
   <td style="text-align:right;"> 818.5050 </td>
   <td style="text-align:right;"> 166.4905 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> flipper_length_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 198.2269 </td>
   <td style="text-align:right;"> 198.2269 </td>
   <td style="text-align:right;"> 40.3210 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> body_mass_g </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 160.3760 </td>
   <td style="text-align:right;"> 160.3760 </td>
   <td style="text-align:right;"> 32.6218 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> sex </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 133.7191 </td>
   <td style="text-align:right;"> 133.7191 </td>
   <td style="text-align:right;"> 27.1995 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Residuals </td>
   <td style="text-align:right;"> 326 </td>
   <td style="text-align:right;"> 1602.6899 </td>
   <td style="text-align:right;"> 4.9162 </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
  </tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
</div>
<p>By default, R breaks down the model’s sum of squares (SSR) into the individual explanatory variables using <em>sequential</em> (also called conditional) sums of squares. A sequential sum of squares describes how much variation an explanatory variable contributes given that the model already includes other explanatory variables.</p>
<p>The order presented in <a href="#tbl-anova-example-R" class="quarto-xref">Table&nbsp;<span>18.3</span></a> is the order in which variables are added to the model. For example, the second row shows <span class="math inline">\(SS(\text{bill\_depth\_mm} | \text{species})\)</span>, meaning bill depth contributes 818.505 additional unique explained variation in the response variable that species has not already explained. The third row shows <span class="math inline">\(SS(\text{flipper\_length\_mm} | \text{species}, \text{bill\_depth\_mm})\)</span>, i.e., how much additional unique variation flipper length explains in a model that includes bill depth and species.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Mathematically, the sequential sum of squares is calculated as the difference in either SSE or SSR between two models—one without the added variable and one with it included. Suppose we want to add variable <span class="math inline">\(X^*\)</span> to a model that already has <span class="math inline">\(k\)</span> other variables, then the calculation is:</p>
<p><span id="eq-seq-ss"><span class="math display">\[
\begin{aligned}
SS(X^*|X_1, \ldots, X_k) &amp;= SSE_{X_1, \ldots, X_k} - SSE_{X_1, \ldots, X_k, X^*} = \\
&amp;= SSR_{X_1, \ldots, X_k, X^*} - SSR_{X_1, \ldots, X_k}
\end{aligned}
\tag{18.4}\]</span></span></p>
<p>Note that SSR increases with each additional variable added to the model, while SSE always decreases. Variation must always be positive, hence we calculate <span class="math inline">\(SSE_{reduced} - SSE_{full}\)</span> or <span class="math inline">\(SSR_{full} - SSR_{reduced}\)</span>.</p>
</div>
</div>
<p>Sequential sums of squares are affected by the order in which variables are added to the model. Let’s change the order of the explanatory variables when estimating the model:</p>
<div class="cell" data-tbl-cap-location="top">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> bill_length_mm <span class="sc">~</span> sex <span class="sc">+</span> ., <span class="at">data =</span> modelData)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(model) <span class="sc">|&gt;</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">4</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-anova-example-sex" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-anova-example-sex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;18.4: Different order of model variables
</figcaption>
<div aria-describedby="tbl-anova-example-sex-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">  </th>
   <th style="text-align:right;"> Df </th>
   <th style="text-align:right;"> Sum Sq </th>
   <th style="text-align:right;"> Mean Sq </th>
   <th style="text-align:right;"> F value </th>
   <th style="text-align:right;"> Pr(&gt;F) </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> sex </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 1175.4780 </td>
   <td style="text-align:right;"> 1175.4780 </td>
   <td style="text-align:right;"> 239.1017 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> species </td>
   <td style="text-align:right;"> 2 </td>
   <td style="text-align:right;"> 6975.5916 </td>
   <td style="text-align:right;"> 3487.7958 </td>
   <td style="text-align:right;"> 709.4457 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> bill_depth_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 64.4987 </td>
   <td style="text-align:right;"> 64.4987 </td>
   <td style="text-align:right;"> 13.1196 </td>
   <td style="text-align:right;"> 0.0003 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> flipper_length_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 78.3815 </td>
   <td style="text-align:right;"> 78.3815 </td>
   <td style="text-align:right;"> 15.9434 </td>
   <td style="text-align:right;"> 0.0001 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> body_mass_g </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 32.2629 </td>
   <td style="text-align:right;"> 32.2629 </td>
   <td style="text-align:right;"> 6.5625 </td>
   <td style="text-align:right;"> 0.0109 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Residuals </td>
   <td style="text-align:right;"> 326 </td>
   <td style="text-align:right;"> 1602.6899 </td>
   <td style="text-align:right;"> 4.9162 </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
  </tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
</div>
<p>In <a href="#tbl-anova-example-sex" class="quarto-xref">Table&nbsp;<span>18.4</span></a> we see that <span class="math inline">\(SS(\text{sex}) = 1175.478\)</span>, which is significantly higher than <span class="math inline">\(SS(\text{sex}|\text{species}, \text{bill\_depth\_mm}, \text{flipper\_length\_mm}, \text{body\_mass\_g}) = 133.7191\)</span> from <a href="#tbl-anova-example-R" class="quarto-xref">Table&nbsp;<span>18.3</span></a>. The variable sex contributes a lot of variation when it is alone in a model, but when added to a model that already includes other variables, it does not contribute as much unique information. This means that the variation explained by the variable also appears to be present in the other variables. We will revisit this observation in a later chapter.</p>
<p>One thing that remains the same in both tables is SSE. We have included the same variables in both models, which means SSY, SSR, and SSE are overall the same. The sum of all sequential sums of squares should still equal SSR regardless of the order of variables, and due to the additive property of variation, SST and SSE have not changed either.</p>
</section>
<section id="partial-f-test-for-groups-of-parameters" class="level4" data-number="18.1.7.2">
<h4 data-number="18.1.7.2" class="anchored" data-anchor-id="partial-f-test-for-groups-of-parameters"><span class="header-section-number">18.1.7.2</span> Partial F-test for groups of parameters</h4>
<p>Sometimes we are interested in examining parts of the model, e.g.&nbsp;a group of slope parameters, whether several variables together contribute explained variation to the model. Instead of examining all slope parameters, we now test a subset which is reflected in the hypothesis:</p>
<p><span class="math display">\[\begin{align*}
H_0&amp;: \beta_1 = \beta_2 = \beta_3 = \cdots = \beta_s = 0\\
H_a&amp;: \text{At least one of } \beta_j \text{ in } H_0 \text{ is different from } 0
\end{align*}\]</span></p>
<p>where <span class="math inline">\(s\)</span> is the number of parameters being tested.</p>
<p>The test statistic for a partial F-test requires a full model (denoted <span class="math inline">\(_F\)</span>) and a reduced model (denoted <span class="math inline">\(_R\)</span>). The full model includes all variables, while the reduced model assumes <span class="math inline">\(H_0\)</span> is true and excludes the variables being tested. We can use either SSR or SSE to calculate how much explained variation is lost between the two models, following the same principle as <a href="#eq-seq-ss" class="quarto-xref">Equation&nbsp;<span>18.4</span></a>.</p>
<p><span id="eq-partial-f"><span class="math display">\[
F_{test} = \frac{(SSR_F - SSR_R) / s}{SSE_F / (n - (k+1))} = \frac{(SSE_R - SSE_F) / s}{SSE_F / (n - (k+1))}
\tag{18.5}\]</span></span></p>
<p>The test statistic still follows an F-distribution with <span class="math inline">\(s\)</span> and <span class="math inline">\(n - (k+1)\)</span> degrees of freedom.</p>
<p>Using <a href="#eq-seq-ss" class="quarto-xref">Equation&nbsp;<span>18.4</span></a>, <a href="#eq-partial-f" class="quarto-xref">Equation&nbsp;<span>18.5</span></a> can be reformulated in a third way that simplifies our analysis process. We can rewrite the difference in explained variation between the full and reduced models as a sequential sum of squares. For example, we may want to investigate whether the variable species is associated with the response variable. Since that variable is transformed into two indicator variables, the hypotheses involve two slope parameters.</p>
<p><span class="math display">\[\begin{align*}
H_0&amp;: \beta_{Chinstrap} = \beta_{Gentoo} = 0\\
H_a&amp;: \text{At least one of } \beta_j \text{ in } H_0 \text{ is different from } 0
\end{align*}\]</span></p>
<p>The reduced model is created assuming <span class="math inline">\(H_0\)</span> is true, i.e., <span class="math inline">\(\beta_{Chinstrap} = \beta_{Gentoo} = 0\)</span>, and the explained variation for the two models is denoted:</p>
<p><span class="math display">\[
\begin{aligned}
  SSR_{R} &amp;= SSR_{bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g, sex} \\
  SSR_{F} &amp;= SSR_{bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g, sex, species}
\end{aligned}
\]</span></p>
<p>We can rewrite the numerator in <a href="#eq-partial-f" class="quarto-xref">Equation&nbsp;<span>18.5</span></a> as:</p>
<p><span class="math display">\[
SS(species|bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g, sex)
\]</span></p>
<p>In the ANOVA tables presented earlier, we can obtain this sum of squares directly if species is added as the last variable in the model.</p>
<div class="cell" data-tbl-cap-location="top">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> bill_depth_mm <span class="sc">+</span> flipper_length_mm <span class="sc">+</span> body_mass_g <span class="sc">+</span> sex <span class="sc">+</span> species, <span class="at">data =</span> modelData)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(model) <span class="sc">|&gt;</span> </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">4</span>) <span class="sc">|&gt;</span> </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>() <span class="sc">|&gt;</span> </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-anova-example-species" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-anova-example-species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;18.5: ANOVA table from a model where species is added last
</figcaption>
<div aria-describedby="tbl-anova-example-species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">  </th>
   <th style="text-align:right;"> Df </th>
   <th style="text-align:right;"> Sum Sq </th>
   <th style="text-align:right;"> Mean Sq </th>
   <th style="text-align:right;"> F value </th>
   <th style="text-align:right;"> Pr(&gt;F) </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> bill_depth_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 518.9806 </td>
   <td style="text-align:right;"> 518.9806 </td>
   <td style="text-align:right;"> 105.5648 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> flipper_length_mm </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 4045.7248 </td>
   <td style="text-align:right;"> 4045.7248 </td>
   <td style="text-align:right;"> 822.9329 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> body_mass_g </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 6.1329 </td>
   <td style="text-align:right;"> 6.1329 </td>
   <td style="text-align:right;"> 1.2475 </td>
   <td style="text-align:right;"> 0.2649 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> sex </td>
   <td style="text-align:right;"> 1 </td>
   <td style="text-align:right;"> 68.4245 </td>
   <td style="text-align:right;"> 68.4245 </td>
   <td style="text-align:right;"> 13.9181 </td>
   <td style="text-align:right;"> 0.0002 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> species </td>
   <td style="text-align:right;"> 2 </td>
   <td style="text-align:right;"> 3686.9500 </td>
   <td style="text-align:right;"> 1843.4750 </td>
   <td style="text-align:right;"> 374.9776 </td>
   <td style="text-align:right;"> 0.0000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Residuals </td>
   <td style="text-align:right;"> 326 </td>
   <td style="text-align:right;"> 1602.6899 </td>
   <td style="text-align:right;"> 4.9162 </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
  </tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
</div>
<p>An ANOVA table with sequential sums of squares calculates a partial F-test for each variable (and its parameter(s)), examining whether the variable contributes a significant increase in explained variation to a model that already includes the variables above. <a href="#tbl-anova-example-species" class="quarto-xref">Table&nbsp;<span>18.5</span></a> now calculates the partial F-test for species (<span class="math inline">\(F_{test} = 374.9776\)</span>), which we were interested in, and we can directly interpret the p-value for the test (<span class="math inline">\(p-value &lt; 0.001\)</span>) as indicating that at least one of the slope parameters is significantly different from 0.</p>
<p>If we conduct a partial F-test for <strong>multiple variables</strong>, we cannot use the p-values shown in the table, since the hypotheses involve more slope parameters/variables than the sequential sums of squares represent. Suppose we want to investigate whether species and sex together contribute to the model. The hypothesis test would then involve:</p>
<p><span class="math display">\[
\begin{aligned}
H_0&amp;: \beta_{sexMale} = \beta_{Chinstrap} = \beta_{Gentoo} = 0\\
H_a&amp;: \text{At least one of } \beta_j \text{ in } H_0 \text{ is different from } 0
\end{aligned}
\]</span></p>
<p>The sequential sum of squares we want to use is denoted <span class="math inline">\(SS(species, sex|bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g)\)</span>, and we can calculate this value by summing the SS for the two variables from <a href="#tbl-anova-example-species" class="quarto-xref">Table&nbsp;<span>18.5</span></a>:</p>
<p><span class="math display">\[
\begin{aligned}
SS(species, sex|bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g) = \\
SS(species|bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g, sex) + \\
SS(sex|bill\_depth\_mm, flipper\_length\_mm, body\_mass\_g)
\end{aligned}
\]</span></p>
<p>Alternatively, we can fit two models in R—the full and reduced and use <code>anova()</code> on both to get a summary output we can interpret directly.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>fullModel <span class="ot">&lt;-</span> simpleModel</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>reducedModel <span class="ot">&lt;-</span> <span class="fu">lm</span>(bill_length_mm <span class="sc">~</span> bill_depth_mm <span class="sc">+</span> flipper_length_mm <span class="sc">+</span> body_mass_g, <span class="at">data =</span> modelData)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(reducedModel, fullModel) <span class="sc">|&gt;</span> </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">4</span>) <span class="sc">|&gt;</span> </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable</span>() <span class="sc">|&gt;</span> </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_styling</span>(<span class="st">"striped"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table quarto-disable-processing="true" class="table table-striped" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:right;"> Res.Df </th>
   <th style="text-align:right;"> RSS </th>
   <th style="text-align:right;"> Df </th>
   <th style="text-align:right;"> Sum of Sq </th>
   <th style="text-align:right;"> F </th>
   <th style="text-align:right;"> Pr(&gt;F) </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:right;"> 329 </td>
   <td style="text-align:right;"> 5358.064 </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
   <td style="text-align:right;">  </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 326 </td>
   <td style="text-align:right;"> 1602.690 </td>
   <td style="text-align:right;"> 3 </td>
   <td style="text-align:right;"> 3755.374 </td>
   <td style="text-align:right;"> 254.6244 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
</tbody>
</table>

</div>
</div>
<p>Since the p-value is less than 0.05, we can conclude that at least one parameter from sex and/or species are relevant to keep in the model, hence we cannot remove them.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>If a parameter is <strong>not</strong> considered significant, it is a justification for removing the variable. We fit a reduced model and begin a new analysis. If a variable is removed, the other parameter estimates will change, and interpretations and inference need to be updated.</p>
</div>
</div>
</section>
</section>
<section id="model-metrics" class="level3" data-number="18.1.8">
<h3 data-number="18.1.8" class="anchored" data-anchor-id="model-metrics"><span class="header-section-number">18.1.8</span> Model metrics</h3>
<p>Because SSR always increases as more variables are added to a model, we need to adjust the metric to compare models of different sizes. We should instead look at the <em>adjusted coefficient of determination</em> (<span class="math inline">\(R^2_{a}\)</span>) to see which model is best. An improved <span class="math inline">\(R^2_{a}\)</span> means the model has removed unnecessary complexity.</p>
<p><span class="math display">\[
R^2_a = 1 - \frac{SSE / (n - (k+1))}{SST / (n - 1)}
\]</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(simpleModel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = bill_length_mm ~ ., data = modelData)

Residuals:
    Min      1Q  Median      3Q     Max 
-7.3939 -1.3424 -0.0421  1.2695 11.4274 

Coefficients:
                   Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)       1.502e+01  4.374e+00   3.433 0.000674 ***
speciesChinstrap  9.566e+00  3.497e-01  27.351  &lt; 2e-16 ***
speciesGentoo     6.404e+00  1.030e+00   6.215 1.56e-09 ***
bill_depth_mm     3.130e-01  1.541e-01   2.032 0.043000 *  
flipper_length_mm 6.856e-02  2.315e-02   2.961 0.003293 ** 
body_mass_g       1.084e-03  4.231e-04   2.562 0.010864 *  
sexmale           2.030e+00  3.892e-01   5.215 3.27e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.217 on 326 degrees of freedom
Multiple R-squared:  0.8386,    Adjusted R-squared:  0.8356 
F-statistic: 282.3 on 6 and 326 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>This can be found in the second to last row in the <code>summary</code> output: <code>Adjusted R-squared:  0.8356</code>. This number cannot be interpreted by itself, but comparing it against the adjusted coefficient for the reduced model (<span class="math inline">\(R^2_a = 0.4554\)</span>), we see that a lot of explained variation has been lost when removing the two categorical variables, adjusted for the model size — another indication that the two variables provide information to the model and should not be removed.</p>
</section>
</section>
<section id="sec-theilsen" class="level2" data-number="18.2">
<h2 data-number="18.2" class="anchored" data-anchor-id="sec-theilsen"><span class="header-section-number">18.2</span> A non-parametric method: Theil–Sen regression</h2>
<p>As we have seen for statistical inference when comparing means or proportions in <a href="23_Statistical_inference.html" class="quarto-xref"><span>Chapter 12</span></a> and <a href="25_Statistical_inference_two.html" class="quarto-xref"><span>Chapter 14</span></a>, there are times when the assumptions for a parametric test or model are not fulfilled. Instead of using a linear regression model that would provide unreliable results, a non-parametric alternative to this method is <em>Theil–Sen regression</em>. This is generally much more robust against outliers than the least-squares method. It also does not require that the residuals are normally distributed. There are also two disadvantages, the main one being that it can only be used for simple regression (one single predictor). It can also be slower to compute, but with today’s computers, this is rarely an issue.</p>
<p>The way Theil–Sen regression works is simple:</p>
<ul>
<li>A line is fit between all possible pairs of points, and their slopes are recorded.</li>
<li>The overall regression slope <em>m</em> is the median of all these pairwise slopes.</li>
<li>The intercept <em>b</em> is the median of all <em>y</em><sub><em>i</em></sub> – <em>m</em> <em>x</em><sub><em>i</em></sub> values, where <em>x</em><sub><em>i</em></sub> is the <em>i</em>th measurement of the predictor and <em>y</em><sub><em>i</em></sub> the corresponding response.</li>
</ul>
<p>To use the Theil–Sen regression, one has to install the package <code>mblm</code> (“median-based linear models”):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"mblm"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mblm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function performing the regression is itself called <code>mblm</code>. A note of caution: its data argument, for some reason, is not called <code>data</code> but <code>dataframe</code>. Let us apply it to set 3 in the Anscombe dataset (the one with the single strong outlier):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mblm</span>(y <span class="sc">~</span> x, <span class="at">dataframe =</span> <span class="fu">filter</span>(ansLong, set <span class="sc">==</span> <span class="st">"3"</span>)) <span class="sc">|&gt;</span> <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
mblm(formula = y ~ x, dataframe = filter(ansLong, set == "3"))

Residuals:
    Min      1Q  Median      3Q     Max 
-0.0045 -0.0022  0.0000  0.0025  4.2435 

Coefficients:
             Estimate       MAD V value Pr(&gt;|V|)   
(Intercept) 4.0050000 0.0074130      65  0.00501 **
x           0.3455000 0.0007413      66  0.00380 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.415 on 9 degrees of freedom</code></pre>
</div>
</div>
<p>As seen, the predicted intercept and slope are no longer 3 and 0.5, but 4 and 0.35 instead. Also, in the regression table above, the median absolute deviation (MAD) of the parameters is reported instead of their standard error, as the MAD is a non-parametric measure of spread.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> The p-values (<code>Pr(&gt;|V|)</code>) in the table are like those in the regression tables from applying <code>summary</code> to <code>lm</code>—however, since Theil–Sen regression is a non-parametric method, these p-values are based on a Wilcoxon rank sum test instead of a t-test.</p>
<p>We can visualize the Theil–Sen regression alongside the least-squares regression, for a better comparison of what they do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>leastSquaresFit <span class="ot">&lt;-</span> <span class="fu">lm</span>(y <span class="sc">~</span> x, <span class="at">data =</span> <span class="fu">filter</span>(ansLong, set <span class="sc">==</span> <span class="st">"3"</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>TheilSenFit <span class="ot">&lt;-</span> <span class="fu">mblm</span>(y <span class="sc">~</span> x, <span class="at">dataframe =</span> <span class="fu">filter</span>(ansLong, set <span class="sc">==</span> <span class="st">"3"</span>))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>ansLong <span class="sc">|&gt;</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(set <span class="sc">==</span> <span class="st">"3"</span>) <span class="sc">|&gt;</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">least squares</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">predict</span>(leastSquaresFit),</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>         <span class="st">`</span><span class="at">Theil-Sen</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">predict</span>(TheilSenFit)) <span class="sc">|&gt;</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"least squares"</span>, <span class="st">"Theil-Sen"</span>),</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"type"</span>, <span class="at">values_to =</span> <span class="st">"prediction"</span>) <span class="sc">|&gt;</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y), <span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> prediction), <span class="at">color =</span> <span class="st">"goldenrod"</span>) <span class="sc">+</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(. <span class="sc">~</span> type) <span class="sc">+</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="42_Multiple_and_Nonlinear_regression_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The Theil–Sen regression correctly recognizes the outlier for what it is, and remains unaffected by it.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the code above, we relied on a function called <code>predict</code>. This simply returns the model-predicted results for each value of the predictor:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(leastSquaresFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       1        2        3        4        5        6        7        8 
7.999727 7.000273 9.498909 7.500000 8.499455 9.998636 6.000818 5.001364 
       9       10       11 
8.999182 6.500545 5.501091 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(TheilSenFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     1      2      3      4      5      6      7      8      9     10     11 
7.4600 6.7690 8.4965 7.1145 7.8055 8.8420 6.0780 5.3870 8.1510 6.4235 5.7325 </code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="sec-nonlin-regression" class="level2" data-number="18.3">
<h2 data-number="18.3" class="anchored" data-anchor-id="sec-nonlin-regression"><span class="header-section-number">18.3</span> Nonlinear regression</h2>
<p>Let us look at another more complex type of model, one where the assumption of linearity does not hold. The built-in <code>CO2</code> data frame contains measurements from an experiment on the cold tolerance of the grass species <a href="https://en.wikipedia.org/wiki/Echinochloa_crus-galli"><em>Echinochloa crus-galli</em></a>. The dataset has five columns:</p>
<ul>
<li><code>Plant</code>: unique identifier for each plant individual.</li>
<li><code>Type</code>: either <code>Quebec</code> or <code>Mississippi</code>, depending on the origin of the plant.</li>
<li><code>Treatment</code>: whether the plant individual was <code>chilled</code> or <code>nonchilled</code> for the experiment.</li>
<li><code>conc</code>: carbon dioxide concentration in the surrounding environment.</li>
<li><code>uptake</code>: carbon dioxide uptake rate.</li>
</ul>
<p>During the exploratory stage of the analysis we discover this relationship between the variables, concentration and uptake separated by each treatment and plant origin:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>CO2 <span class="sc">|&gt;</span> </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> uptake) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span> </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(Type <span class="sc">~</span> Treatment) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-non-lin-co2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-non-lin-co2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-non-lin-co2-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-non-lin-co2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.11: Example of non-linear relationships
</figcaption>
</figure>
</div>
</div>
</div>
<p>The relationship between CO<sub>2</sub> concentration and uptake rates are definitely not linear, regardless of treatment or type. So the question arises: can one fit a <em>non-linear</em> function to these data? As an example, let us focus on just Quebec and the non-chilled treatment, to better illustrate the ideas behind nonlinear regression. Here are the filtered data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(CO2) <span class="sc">|&gt;</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Type <span class="sc">==</span> <span class="st">"Quebec"</span>, Treatment <span class="sc">==</span> <span class="st">"nonchilled"</span>) <span class="sc">|&gt;</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> uptake)) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>) <span class="sc">+</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-co2-filter" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-co2-filter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="42_Multiple_and_Nonlinear_regression_files/figure-html/fig-co2-filter-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-co2-filter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18.12: Relationship between concentration and uptake rate of CO2 in Quebec using the non-chilled treatment
</figcaption>
</figure>
</div>
</div>
</div>
<p>If the function we wish to fit is not linear, we have to specify its shape. One commonly used shape for describing the above saturating pattern is the <em>Michaelis–Menten curve</em>. This is given by the following equation: <span class="math display">\[ \rho = \frac{V c}{K + c} \]</span> Here <span class="math inline">\(\rho\)</span> is the uptake rate, <span class="math inline">\(c\)</span> is the concentration, and <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span> are two parameters which can modify the shape of the function. The figure below illustrates what curves one can get by varying these parameters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">expand_grid</span>(<span class="at">V =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>), <span class="co"># This function creates a tibble with all</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">K =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">10</span>),   <span class="co"># possible combinations of the inputs</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">concentration =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">60</span>, <span class="at">l =</span> <span class="dv">201</span>)) <span class="sc">|&gt;</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(V, K) <span class="sc">|&gt;</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">uptake =</span> V <span class="sc">*</span> concentration <span class="sc">/</span> (K <span class="sc">+</span> concentration)) <span class="sc">|&gt;</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">|&gt;</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> concentration, <span class="at">y =</span> uptake)) <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(V <span class="sc">~</span> K, <span class="at">labeller =</span> label_both) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="42_Multiple_and_Nonlinear_regression_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The task is to find the values of <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span> that provide the best fit to the data. Like in the case of linear regression, this can be done via the least-squares criterion: the best fit is provided by the curve which minimizes the sum of the squared deviations of the observed points from it. Unlike with linear regression however, this curve can be very difficult to find. In fact, there is no known general procedure that would be able to minimize the sum of squares under all circumstances. What algorithms <em>can</em> do is to find the best fit, given some initial guesses for the parameters that are at least not violently off of the true values. Just how close the guess needs to be is context-dependent, and highlights an important problem: non-linear regression can be as much an art as it is a science. For the types of curves we will be fitting though, the more subtle problems will never come up, and a “good enough” initial guess can vary within a relatively wide range.</p>
<p>So, how can one guess the values of <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span>? To do this, one has to have an understanding of how the parameters influence the curves. For <span class="math inline">\(V\)</span>, this interpretation is not difficult to infer. Notice that if concentrations are very, very large, then in the denominator of the formula <span class="math inline">\(\rho = V c / (K + c)\)</span>, we might as well say that <span class="math inline">\(K + c\)</span> is approximately equal to <span class="math inline">\(c\)</span> (if <span class="math inline">\(c\)</span> is a million and <span class="math inline">\(K\)</span> is one, then one is justified in treating the sum as being about one million still). This means that for large <span class="math inline">\(c\)</span>, the formula reduces to <span class="math inline">\(\rho \approx V c / c = V\)</span>. In other words <span class="math inline">\(V\)</span> is the <em>saturation uptake rate:</em> the maximum value of the uptake. This, incidentally, is clearly visible in the plots above: when <span class="math inline">\(V\)</span> is 10 (top row), the curves always tend towards 10 for large concentrations; when <span class="math inline">\(V\)</span> is 20, they tend towards 20 (middle row), and when <span class="math inline">\(V\)</span> is 30, they tend towards 30.</p>
<p>The interpretation of <span class="math inline">\(K\)</span> is slightly less straightforward, but still simple. To see what it means, let us ask what the uptake rate would be, were the concentration’s value equal to <span class="math inline">\(K\)</span>. In that case, we get <span class="math inline">\(\rho = V K / (K + K)\)</span> (we simply substituted <span class="math inline">\(c = K\)</span> into the formula), or <span class="math inline">\(\rho = VK / (2K) = V/2\)</span>. That is, <span class="math inline">\(K\)</span> is the concentration at which the uptake rate reaches half its maximum.</p>
<p>Looking at the data again, both these parameters can be roughly estimated:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(CO2) <span class="sc">|&gt;</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Type <span class="sc">==</span> <span class="st">"Quebec"</span>, Treatment <span class="sc">==</span> <span class="st">"nonchilled"</span>) <span class="sc">|&gt;</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> uptake)) <span class="sc">+</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">yintercept =</span> <span class="dv">43</span>, </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"steelblue"</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">geom =</span> <span class="st">"segment"</span>, </span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">43</span><span class="sc">/</span><span class="dv">2</span>, </span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">xend =</span> <span class="dv">125</span>, <span class="at">yend =</span> <span class="dv">43</span><span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"steelblue"</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">geom =</span> <span class="st">"segment"</span>, </span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="dv">125</span>, <span class="at">y =</span> <span class="dv">43</span><span class="sc">/</span><span class="dv">2</span>, </span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">xend =</span> <span class="dv">125</span>, <span class="at">yend =</span> <span class="dv">0</span>,</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"steelblue"</span></span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>) <span class="sc">+</span></span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="42_Multiple_and_Nonlinear_regression_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>So guessing that <span class="math inline">\(V\)</span> is about 43 and <span class="math inline">\(K\)</span> is about 125 seems to be close to the mark.</p>
<p>To actually perform the nonlinear regression, one can use the <code>nls</code> function (“Nonlinear Least Squares”). It begins much like <code>lm</code>, taking a formula and a data frame. However, the formula is no longer a shorthand notation for a linear model, and therefore has to be entered literally. Additionally, there is another argument to <code>nls</code> called <code>start</code>; this is where the starting values have to be specified. The <code>start</code> argument has to be in the form of a <em>list</em>. Lists are an important data structure, worth a little interlude to explain how they work.</p>
<section id="sec-lists" class="level3" data-number="18.3.1">
<h3 data-number="18.3.1" class="anchored" data-anchor-id="sec-lists"><span class="header-section-number">18.3.1</span> Interlude: lists</h3>
<p>Lists are like vectors except they can hold arbitrary data in their entries. So unlike vectors which are composed of either all numbers or all character strings or all logical values, lists may have a combination of these. Furthermore, list entries are not restricted to elementary types: vectors, or even data frames may also be list entries. To define a list, all one needs to do is type <code>list</code>, and then give a sequence of named entries. For example, the following creates a list with three entries: <code>a</code> will be equal to 3, <code>b</code> to the string <code>"Hello!"</code>, and <code>myTable</code> to a small tibble.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list</span>(<span class="at">a =</span> <span class="dv">3</span>, <span class="at">b =</span> <span class="st">"Hello!"</span>, <span class="at">myTable =</span> <span class="fu">tibble</span>(<span class="at">x =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">y =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$a
[1] 3

$b
[1] "Hello!"

$myTable
# A tibble: 2 × 2
      x     y
  &lt;dbl&gt; &lt;dbl&gt;
1     1     3
2     2     4</code></pre>
</div>
</div>
<p>One can refer to the entries of the list either with the <code>$</code> notation, or using double brackets. Assigning the above list to a variable called <code>myList</code> first:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>myList <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">a =</span> <span class="dv">3</span>,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">b =</span> <span class="st">"Hello!"</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">myTable =</span> <span class="fu">tibble</span>(<span class="at">x =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">y =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>))</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now access the entries of <code>myList</code> either as</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>myList<span class="sc">$</span>myTable <span class="co"># Access the entry called myTable in the list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
      x     y
  &lt;dbl&gt; &lt;dbl&gt;
1     1     3
2     2     4</code></pre>
</div>
</div>
<p>Or as</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>myList[[<span class="dv">3</span>]] <span class="co"># Access the third entry (myTable) in the list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
      x     y
  &lt;dbl&gt; &lt;dbl&gt;
1     1     3
2     2     4</code></pre>
</div>
</div>
<p>The <code>$</code> notation is the same as when one accesses columns of data frames. This is not a coincidence: internally, data frames are represented as lists of columns, with each column being a vector. It follows that the double-bracket notation can also be used in conjunction with data frames: to access the first column of <code>CO2</code>, we can not only write <code>CO2$Plant</code>, but also <code>CO2[[1]]</code>.</p>
<p>A list seems to be just a more flexible version of a vector—so why would we want to use vectors instead of lists in the first place? The answer has to do with efficiency: the price to pay for the increased flexibility offered by lists is that they are slower to do operations on. While this would not be a problem for the applications found in this book, it can become important when dealing with large datasets or heavy numerical computations. As a corollary, R has many useful functions which work on vectors but do not work by default on lists. To mention just the simplest examples: <code>mean</code>, <code>median</code>, <code>sd</code>, and <code>sum</code> will throw an error when applied to lists. That is,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>returns the expected 6, because it was applied to a vector. But the same expression results in an error when the vector is replaced by a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">list</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in sum(list(1, 2, 3)): invalid 'type' (list) of argument</code></pre>
</div>
</div>
</section>
<section id="back-to-nonlinear-regression" class="level3" data-number="18.3.2">
<h3 data-number="18.3.2" class="anchored" data-anchor-id="back-to-nonlinear-regression"><span class="header-section-number">18.3.2</span> Back to nonlinear regression</h3>
<p>With this brief introduction to lists, we now understand what it means that the <code>start</code> argument to <code>nls</code> must be a list, with the entries named after the parameters to be fitted. Using the previously-guessed values of <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span> being around 43 and 125, respectively, means we can use <code>start = list(V = 43, K = 125)</code>. We can now look at the <code>nls</code> function and use it to produce a fit of the Michaelis–Menten curve to our data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>nonlinearFit <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(CO2) <span class="sc">|&gt;</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Type <span class="sc">==</span> <span class="st">"Quebec"</span>, Treatment <span class="sc">==</span> <span class="st">"nonchilled"</span>) <span class="sc">|&gt;</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nls</span>(uptake <span class="sc">~</span> V<span class="sc">*</span>conc<span class="sc">/</span>(K <span class="sc">+</span> conc), <span class="at">data =</span> _, <span class="at">start =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">43</span>, <span class="at">K =</span> <span class="dv">125</span>))</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(nonlinearFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Nonlinear regression model
  model: uptake ~ V * conc/(K + conc)
   data: filter(as_tibble(CO2), Type == "Quebec", Treatment == "nonchilled")
     V      K 
 51.36 136.32 
 residual sum-of-squares: 319.2

Number of iterations to convergence: 6 
Achieved convergence tolerance: 9.722e-06</code></pre>
</div>
</div>
<p>Observe that in the formula, we use the column name <code>conc</code> when we want to use the data inside that column, but use made-up names (in this case, <code>V</code> and <code>K</code>) for the unknown parameters we are trying to obtain. Their starting values were filled in from our earlier visual estimation. From these starting values, the best fitting solution is found. We see that their values are 51.36 for <span class="math inline">\(V\)</span> and 136.32 for <span class="math inline">\(K\)</span>.</p>
<p>The result of <code>nls</code> can be used inside <code>summary</code> to get more information on the fitted parameters (note: the <code>anova</code> function is not applicable to nonlinear regression). Doing so results in the following regression table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(nonlinearFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Formula: uptake ~ V * conc/(K + conc)

Parameters:
  Estimate Std. Error t value Pr(&gt;|t|)    
V   51.359      2.831  18.140 1.86e-13 ***
K  136.319     27.027   5.044 7.21e-05 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.099 on 19 degrees of freedom

Number of iterations to convergence: 6 
Achieved convergence tolerance: 9.722e-06</code></pre>
</div>
</div>
<p>As in the case of linear regression, the statistical analysis will only be reliable if the assumptions of the independence, normality, and homoscedasticity of the residuals are maintained.</p>
<p>To conclude this section, it can be useful to plot the data together with the fitted nonlinear curve, to make sure visually that the fit is reasonable. There are several possible ways of doing this; here we discuss two of them. First, one can rely on the <code>predict</code> function (<a href="#sec-theilsen" class="quarto-xref"><span>Section 18.2</span></a>) that will tell us, for each value of the predictor, what the model-predicted values are. In the same way as for models generated by <code>lm</code> or <code>mblm</code>,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(nonlinearFit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 21.09265 28.87030 33.23631 36.96287 40.35655 42.72982 45.19795 21.09265
 [9] 28.87030 33.23631 36.96287 40.35655 42.72982 45.19795 21.09265 28.87030
[17] 33.23631 36.96287 40.35655 42.72982 45.19795</code></pre>
</div>
</div>
<p>returns a vector for each <code>conc</code> in the original data, documenting what the model thinks the corresponding uptake rate ought to be. They can then be compared with the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(CO2) <span class="sc">|&gt;</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Type <span class="sc">==</span> <span class="st">"Quebec"</span>, Treatment <span class="sc">==</span> <span class="st">"nonchilled"</span>) <span class="sc">|&gt;</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pred =</span> <span class="fu">predict</span>(nonlinearFit)) <span class="sc">|&gt;</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> uptake), <span class="at">color =</span> <span class="st">"steelblue"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> pred), <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>) <span class="sc">+</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="42_Multiple_and_Nonlinear_regression_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In the above plot, the data points and the predictions each had their own set of aeshetics. For this reason, the aesthetic mappings were not defined separately, but locally inside each <code>geom_</code>. This is perfectly legal, and can help whenever different geometries take different aesthetics from the data. Second, notice that the prediction was drawn with a dashed, semi-transparent line. This is intentional, to make it distinct from data. It signals that the curve does not correspond to data we are plotting, but to a model’s predictions.</p>
<p>The second method can be useful if the data are sufficiently scarce that the fitted line looks “rugged”, a bit too piecewise (this can be seen in the above example as well if one looks carefully). In that case, it is possible to draw the curve of any function using <code>geom_function</code>. We can use the fitted parameters in drawing it, and it will not suffer from being too piecewise:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>V_fitted <span class="ot">&lt;-</span> <span class="fu">coef</span>(nonlinearFit)[<span class="st">"V"</span>] <span class="co"># Get fitted values of V and K</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>K_fitted <span class="ot">&lt;-</span> <span class="fu">coef</span>(nonlinearFit)[<span class="st">"K"</span>] <span class="co"># from the vector of coefficients</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(CO2) <span class="sc">|&gt;</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Type <span class="sc">==</span> <span class="st">"Quebec"</span>, Treatment <span class="sc">==</span> <span class="st">"nonchilled"</span>) <span class="sc">|&gt;</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> conc, <span class="at">y =</span> uptake), <span class="at">color =</span> <span class="st">"steelblue"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_function</span>(<span class="at">fun =</span> <span class="cf">function</span>(x) V_fitted <span class="sc">*</span> x <span class="sc">/</span> (K_fitted <span class="sc">+</span> x),</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Concentration"</span>, <span class="at">y =</span> <span class="st">"Uptake Rate"</span>) <span class="sc">+</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="42_Multiple_and_Nonlinear_regression_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The result is much the same as before, although carefully looking at the dashed line shows that the second curve is smoother than the first. In this case, this does not matter much, but it could be aesthetically more relevant in others.</p>
</section>
</section>
<section id="exercises" class="level2" data-number="18.4">
<h2 data-number="18.4" class="anchored" data-anchor-id="exercises"><span class="header-section-number">18.4</span> Exercises</h2>
<p>The following exercises focus on all three of the advanced topics of regression.</p>
<section id="multiple-linear-regression" class="level3" data-number="18.4.1">
<h3 data-number="18.4.1" class="anchored" data-anchor-id="multiple-linear-regression"><span class="header-section-number">18.4.1</span> Multiple linear regression</h3>
<p>Use the <code>iris</code> dataset from <a href="05_Basic_data_wrangling.html#sec-wrangling-exercises" class="quarto-xref"><span>Section 5.3</span></a> and complete a regression analysis outlined below focusing on a model describing <code>petal_length</code>.</p>
<ol type="1">
<li>Investigate the type and scale of the variables.</li>
<li>Summarize descriptive statistics for each variable.</li>
<li>Visualize the distribution of each variable.</li>
<li>Create a scatter plot showing the relationship between the two variables in the sample. Interpret it based on the four key pieces of information a scatter plot can provide.</li>
<li>Summarize your observations and justify whether or not you think a simple linear regression model is suitable.</li>
<li>Fit a linear regression model using <code>lm()</code> and interpret the regression coefficients.</li>
<li>Create diagnostic plots to check whether the assumptions of the model are satisfied.</li>
<li>Perform an F-test for the model and interpret the result. Do the same for a t-test for the slope parameter.</li>
<li>Summarize your findings along with the coefficient of determination and determine whether or not the model is suitable for predicting plant growth based on soil moisture content.</li>
</ol>
</section>
<section id="non-parametric-regression" class="level3" data-number="18.4.2">
<h3 data-number="18.4.2" class="anchored" data-anchor-id="non-parametric-regression"><span class="header-section-number">18.4.2</span> Non-parametric regression</h3>
<p>Use the plant growth rate data set from <a href="41_Linear_regression.html#sec-exercises-regression" class="quarto-xref"><span>Section 17.7</span></a> and estimate a Theil-Sen regression. How does the relationship differ between the two models?</p>
</section>
<section id="sec-exercises-non-linear" class="level3" data-number="18.4.3">
<h3 data-number="18.4.3" class="anchored" data-anchor-id="sec-exercises-non-linear"><span class="header-section-number">18.4.3</span> Non-linear regression</h3>
<section id="exponential-growth" class="level4" data-number="18.4.3.1">
<h4 data-number="18.4.3.1" class="anchored" data-anchor-id="exponential-growth"><span class="header-section-number">18.4.3.1</span> Exponential growth</h4>
<p>Let <span class="math inline">\(N(t)\)</span> be the population abundance of some organism at time <span class="math inline">\(t\)</span>. An exponentially growing population increases according to <span class="math display">\[ N(t) = N_0 \mathrm{e}^{rt} \]</span> where <span class="math inline">\(N_0\)</span> is the initial population size at time <span class="math inline">\(t=0\)</span>, and <span class="math inline">\(r\)</span> is the exponential rate of increase.</p>
<ol type="1">
<li>Download the data file <a href="https://raw.githubusercontent.com/dysordys/intro-R-and-stats/main/data/pop_growth_1.zip"><code>pop_growth_1.csv</code></a> and load it in R.</li>
<li>Use <code>nls()</code> to fit the above exponential growth model to this dataset. Do not treat <span class="math inline">\(N_0\)</span> as a free parameter, but instead use the actual population size at time <span class="math inline">\(t=0\)</span>. This leaves <span class="math inline">\(r\)</span> as the only parameter to be fitted. Do so, using an appropriate starting value.</li>
<li>Assume that the data describes a population of water lilies, and that a single ‘individual’ weighs 1 gram. If the population would continue to grow unrestricted, what would be its biomass after nine months (about 280 days)? What object would have a comparable weight to this population, and what does that tell us about unrestricted population growth in general?</li>
</ol>
</section>
<section id="nitrogen-uptake" class="level4" data-number="18.4.3.2">
<h4 data-number="18.4.3.2" class="anchored" data-anchor-id="nitrogen-uptake"><span class="header-section-number">18.4.3.2</span> Nitrogen uptake</h4>
<p><span class="citation" data-cites="CedergreenMadsen2002">Cedergreen and Madsen (<a href="99_References.html#ref-CedergreenMadsen2002" role="doc-biblioref">2002</a>)</span> reported data from an experiment on nitrogen uptake by the duckweed <em>Lemna minor</em>, where the predictor variable is the initial substrate concentration and the response variable is the uptake rate. In this type of experiment, it is anticipated that the uptake will increase as the concentration increases, approaching a horizontal asymptote. The data are available in <a href="https://raw.githubusercontent.com/dysordys/intro-R-and-stats/main/data/uptake.zip"><code>uptake.csv</code></a>.</p>
<ol type="1">
<li>Create a plot of the data, with the nitrogen concentrations along the x-axis and the corresponding uptake rates along the y-axis.</li>
<li>Fit a Michaelis-Menten model (describing saturating dynamics) to the data. This model is given by <span class="math display">\[ \rho = \frac{V c}{K + c} \]</span> where <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span> are two constants, <span class="math inline">\(c\)</span> is the concentration, and <span class="math inline">\(\rho\)</span> the uptake rate. Make initial guesses for the two parameters <span class="math inline">\(V\)</span> and <span class="math inline">\(K\)</span> based on the graph, and perform the nonlinear regression.</li>
<li>Given your nonlinear regression results, what is the maximum possible nitrogen uptake rate of <em>L. minor</em>?</li>
</ol>
</section>
<section id="logistic-growth" class="level4" data-number="18.4.3.3">
<h4 data-number="18.4.3.3" class="anchored" data-anchor-id="logistic-growth"><span class="header-section-number">18.4.3.3</span> Logistic growth</h4>
<p>The simplest model illustrating population regulation and regulated growth is the <em>logistic model</em>, defined by the differential equation <span class="math display">\[ \frac{\mathrm{d} N(t)}{\mathrm{d} t} = rN(t) \left( 1 - \frac{N(t)}{K} \right) \]</span> Here <span class="math inline">\(N(t)\)</span> is the population abundance at time <span class="math inline">\(t\)</span>, <span class="math inline">\(r\)</span> is the exponential growth rate of the population when rare, and <span class="math inline">\(K\)</span> is the maximum abundance it can sustainably achieve (the “carrying capacity”). It should be obvious that when <span class="math inline">\(N(t) = K\)</span>, the derivative vanishes, signalling that the population size no longer changes.</p>
<p>The above differential equation is one of the few which can be solved explicitly. Its solution reads <span class="math display">\[ N(t) = N_0 \frac{\mathrm{e}^{rt}}{1-(1-\mathrm{e}^{rt})(N_0/K)} \]</span> where <span class="math inline">\(N_0\)</span> is the initial population size at time <span class="math inline">\(t=0\)</span>. Let us fit this model to some population growth data.</p>
<ol type="1">
<li>Download the data file <a href="https://raw.githubusercontent.com/dysordys/intro-R-and-stats/main/data/pop_growth_2.zip"><code>pop_growth_2.csv</code></a>, load it in R, and plot the population abundances against time.</li>
<li>Fit the above model to the data using the <code>nls()</code> function, with appropriate guesses for the starting values of <span class="math inline">\(r\)</span> and <span class="math inline">\(K\)</span>.</li>
<li>Plot the data and the model prediction together. What are the estimated values of <span class="math inline">\(r\)</span> and <span class="math inline">\(K\)</span>?</li>
</ol>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-CedergreenMadsen2002" class="csl-entry" role="listitem">
Cedergreen, Nina, and Tom Vindbæk Madsen. 2002. <span>“<span class="nocase">Nitrogen uptake by the floating macrophyte Lemna minor</span>.”</span> <em>New Phytologist</em> 155 (2): 285–92. <a href="https://doi.org/10.1046/j.1469-8137.2002.00463.x">https://doi.org/10.1046/j.1469-8137.2002.00463.x</a>.
</div>
<div id="ref-penguins2020" class="csl-entry" role="listitem">
Horst, Allison Marie, Alison Presmanes Hill, and Kristen B Gorman. 2020. <em>Palmerpenguins: Palmer Archipelago (Antarctica) Penguin Data</em>. <a href="https://doi.org/10.5281/zenodo.3960218">https://doi.org/10.5281/zenodo.3960218</a>.
</div>
</div>
</section>
</section>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>The median absolute deviation (MAD) over a set of data points <span class="math inline">\(x_i\)</span> is defined as <span class="math inline">\(\text{MAD} = \text{median}(|x_i - \text{median}(x)|)\)</span>, where <span class="math inline">\(\text{median}(x)\)</span> is the median of the data.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./41_Linear_regression.html" class="pagination-link" aria-label="Simple linear regression">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Simple linear regression</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./43_Anova.html" class="pagination-link" aria-label="Analysis of Variance">
        <span class="nav-page-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>