<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>19&nbsp; Analysis of Variance – Introduction to data analysis and visualization with R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./44_Anova_two_way.html" rel="next">
<link href="./42_Multiple_and_Nonlinear_regression.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-8da5b4427184b79ecddefad3d342027e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<meta name="shinylive:serviceworker_dir" content=".">
<script src="site_libs/quarto-contrib/shinylive-0.9.1/shinylive/load-shinylive-sw.js" type="module"></script>
<script src="site_libs/quarto-contrib/shinylive-0.9.1/shinylive/run-python-blocks.js" type="module"></script>
<link href="site_libs/quarto-contrib/shinylive-0.9.1/shinylive/shinylive.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/shinylive-quarto-css/shinylive-quarto.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./40_Intro_adv_statistics.html">Part III: Advanced statistics</a></li><li class="breadcrumb-item"><a href="./43_Anova.html"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to data analysis and visualization with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Part I: Processing and visualizing data with R</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_Intro_R_RStudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_R_programming_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R programming basics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_Functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_Data_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Reading tabular data from disk</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_Basic_data_wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Basic data manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_Summaries_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Summary statistics and tidy data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_Creating_figures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Creating publication-grade figures</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_Further_plotting_options.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Some further plotting options; introducing factors</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_Joining_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Joining data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Part II: Introductory statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_Probabilities_and_distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Probabilities and probability distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./22_Sampling_distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Sampling distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_Statistical_inference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to statistical inference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./24_Statistical_inference_in_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Statistical inference in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./25_Statistical_inference_two.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Statistical inference in two groups</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./26_Statistical_inference_three.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Statistical inference in qualitative variables</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Part III: Advanced statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_Intro_adv_statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Relationships between variables</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./41_Linear_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Simple linear regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./42_Multiple_and_Nonlinear_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./43_Anova.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./44_Anova_two_way.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Two-way ANOVA and the Scheirer–Ray–Hare test</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./45_Intro_map.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Higher-order functions and mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./46_Multiple_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Nested data and multiple analysis</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">References</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./99_References.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">References</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#one-way-anova" id="toc-one-way-anova" class="nav-link active" data-scroll-target="#one-way-anova"><span class="header-section-number">19.1</span> One-way ANOVA</a>
  <ul>
  <li><a href="#alternative-to-indicator-coding" id="toc-alternative-to-indicator-coding" class="nav-link" data-scroll-target="#alternative-to-indicator-coding"><span class="header-section-number">19.1.1</span> Alternative to indicator coding</a></li>
  </ul></li>
  <li><a href="#the-anova-model" id="toc-the-anova-model" class="nav-link" data-scroll-target="#the-anova-model"><span class="header-section-number">19.2</span> The ANOVA-model</a></li>
  <li><a href="#statistical-inference-in-an-anova-model" id="toc-statistical-inference-in-an-anova-model" class="nav-link" data-scroll-target="#statistical-inference-in-an-anova-model"><span class="header-section-number">19.3</span> Statistical inference in an ANOVA model</a>
  <ul>
  <li><a href="#overall-f-test" id="toc-overall-f-test" class="nav-link" data-scroll-target="#overall-f-test"><span class="header-section-number">19.3.1</span> Overall F-test</a></li>
  <li><a href="#multiple-comparisons" id="toc-multiple-comparisons" class="nav-link" data-scroll-target="#multiple-comparisons"><span class="header-section-number">19.3.2</span> Multiple comparisons</a>
  <ul class="collapse">
  <li><a href="#tukey-family-confidence" id="toc-tukey-family-confidence" class="nav-link" data-scroll-target="#tukey-family-confidence"><span class="header-section-number">19.3.2.1</span> Tukey family confidence</a></li>
  <li><a href="#dunnetts-test-against-control" id="toc-dunnetts-test-against-control" class="nav-link" data-scroll-target="#dunnetts-test-against-control"><span class="header-section-number">19.3.2.2</span> Dunnett’s test against control</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#the-kruskalwallis-and-dunn-tests" id="toc-the-kruskalwallis-and-dunn-tests" class="nav-link" data-scroll-target="#the-kruskalwallis-and-dunn-tests"><span class="header-section-number">19.4</span> The Kruskal–Wallis and Dunn tests</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">19.5</span> Exercises</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./40_Intro_adv_statistics.html">Part III: Advanced statistics</a></li><li class="breadcrumb-item"><a href="./43_Anova.html"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span id="sec-ANOVA" class="quarto-section-identifier"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In the previous chapters we studied the relationship between a continuous response variable and one or more explanatory variables. Within multiple linear regression we allow the explanatory variables to be a mix of continuous and qualitative variables and can fit the regression model using indicator variables.</p>
<p>If a regression model solely contain qualitative explanatory variables, it is equivalent to compare means of different groups of those variables. This can be simplified into an <em>Analysis of Variance</em>, or <em>ANOVA</em>-model, which is often used in experimental studies. Within this family of models, we now call the explanatory variables <em>factors</em> and their different values <em>levels</em>.</p>
<p>Something to keep in mind when reading this chapter is that an ANOVA-model is just a simplified version of a regression model.</p>
<section id="one-way-anova" class="level2" data-number="19.1">
<h2 data-number="19.1" class="anchored" data-anchor-id="one-way-anova"><span class="header-section-number">19.1</span> One-way ANOVA</h2>
<p>An ANOVA-model containing only one factor is called a <em>one-way ANOVA</em>. We can formulate the model mathematically similar to that of a regression model: <span id="eq-indicator-oneway"><span class="math display">\[
  \begin{aligned}
    Y_{ij} = \beta_0 + \beta_1 \cdot X_{1j} + \dots + \beta_{A-1}\cdot X_{(A-1)j} + \varepsilon_{ij}
  \end{aligned}
\tag{19.1}\]</span></span></p>
<p>where</p>
<ul>
<li><span class="math inline">\(j\)</span> is the observation index for the <span class="math inline">\(j\)</span>:th observation in level <span class="math inline">\(i\)</span> of the factor, ranging from 1 to <span class="math inline">\(n_i\)</span>, the number of observations in total within that level</li>
<li><span class="math inline">\(A\)</span> is the number of levels within the factor,</li>
<li>all <span class="math inline">\(X_{ij}\)</span> are indicator coded variables that each indicate level <span class="math inline">\(i\)</span> of the factor.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The difference in an ANOVA-model compared to a standard regression model is the fact we now have two indexes, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>. We will encounter more ANOVA-models with more factors that increase this even further, but one thing to keep in mind is that the last index is considered the observation index within a specific combination of factor levels.</p>
</div>
</div>
<p>This model will in practice predict the same value for all observations of the same level because an indicator variable is the same value for all observations therein. We can simplify the predicted values as:</p>
<p><span class="math display">\[
  \begin{aligned}
    Y_{i = 1} = \mu_1 &amp;= \beta_0 + \beta_1 \cdot 1 + \beta_2 \cdot 0 + \dots + \beta_{A-1} \cdot 0 = \beta_0 + \beta_1\\
    Y_{i = 2} = \mu_2 &amp;= \beta_0 + \beta_1 \cdot 0 + \beta_2 \cdot 1 + \dots + \beta_{A-1} \cdot 0 = \beta_0 + \beta_2\\
    &amp;\quad \vdots \\
    Y_{i = A-1} = \mu_{A-1} &amp;= \beta_0 + \beta_1 \cdot 0 + \beta_2 \cdot 0 + \dots + \beta_{A-1} \cdot 1 = \beta_0 + \beta_{A-1}\\
    Y_{i = A} = \mu_A &amp;= \beta_0 + \beta_1 \cdot 0 + \beta_2 \cdot 0 + \dots + \beta_{A-1} \cdot 0 = \beta_0
  \end{aligned}
\]</span></p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some literature may use another variant of the model, called the <em>cell means</em> model.</p>
<p><span class="math display">\[
  \begin{aligned}
    Y_{ij} = \mu_i + \varepsilon_{ij}
  \end{aligned}
\]</span> where <span class="math inline">\(j\)</span> is the observation index within each level, <span class="math inline">\(\mu_i\)</span> is the mean for level <span class="math inline">\(i\)</span> of the factor.</p>
</div>
</div>
<p>As mentioned multiple times but is worth repeating, an ANOVA-model is a special case of a regression model which means that the model assumptions <span class="math inline">\(\varepsilon_{ij} \overset{\mathrm{iid}}{\sim} N(0, \sigma^2)\)</span> still apply and that the model should be evaluated using residual analysis before moving forward to statistical inference. One of the main differences with an ANOVA-model is that the residuals will be grouped to the different levels of the factor and that it can be a bit tough to assess if the homoscedasticity assumption hold.</p>
<section id="alternative-to-indicator-coding" class="level3" data-number="19.1.1">
<h3 data-number="19.1.1" class="anchored" data-anchor-id="alternative-to-indicator-coding"><span class="header-section-number">19.1.1</span> Alternative to indicator coding</h3>
<p>There exists many other ways to code our indicator variables that simplifies the mathematical calculations when fitting a model. One such way is using a <em>factor effect</em> coding system that has two main benefits; firstly it provides accurate parameter estimates in more complex ANOVA-models, secondly it changes the way we interpret an estimate for a specific level.</p>
<p>Indicator coded variables are interpreted as “the effect of this level compared to the chosen reference level” and the estimated parameters will numerically change based on the, most often arbitrarily, chosen reference level. Factor effect coding instead provides the same numerical parameter estimates regardless of the level that is “skipped” and is interpreted as “the effect of this level compared to the overall mean”.</p>
<p>We can define this new model in its regression form as: <span id="eq-factor-oneway"><span class="math display">\[
  \begin{aligned}
    Y_{ij} = \mu + \beta_1 \cdot X_{1j} + \dots + \beta_{A-1}\cdot X_{(A-1)j} + \varepsilon_{ij}
  \end{aligned}
\tag{19.2}\]</span></span></p>
<p>which is very similar to <a href="#eq-indicator-oneway" class="quarto-xref">Equation&nbsp;<span>19.1</span></a> except that the intercept is now written as <span class="math inline">\(\mu\)</span>. Mathematically the differences are:</p>
<ul>
<li><p><span class="math inline">\(\mu\)</span> is the <strong>unweighted</strong> overall mean of <span class="math inline">\(Y_{ij}\)</span> calculated as: <span class="math display">\[
\begin{aligned}
\mu &amp;= \frac{\mu_1 + \mu_2 + \cdots + \mu_A}{A}
\end{aligned}
\]</span></p></li>
<li><p><span class="math inline">\(\beta_i\)</span> is the difference or effect of level <span class="math inline">\(i\)</span> on the overall mean calculated as:</p>
<p><span class="math display">\[
\begin{aligned}
  \beta_i = \mu_i - \mu
\end{aligned}
\]</span></p></li>
<li><p><span class="math inline">\(X_{ij}\)</span> is factor effect coded as: <span class="math display">\[
\begin{aligned}
  X_{ij} =
  \begin{cases}
      1 \quad &amp;\text{if level i}\\
      -1      &amp;\text{if level A}\\
      0       &amp;\text{otherwise}
  \end{cases}
\end{aligned}
\]</span> With the help of this coding we can show that the effect of level <span class="math inline">\(A\)</span> can be calculated as the negative sum of the other parameters, even though that effect is not specifically defined in the model.</p></li>
</ul>
<p><span class="math display">\[
\begin{aligned}
  \beta_A = -(\beta_1 + \beta_2 + \cdots + \beta_{A-1}) = \mu_A - \mu
\end{aligned}
\]</span></p>
<div id="nte-anova-formulation" class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important&nbsp;19.1
</div>
</div>
<div class="callout-body-container callout-body">
<p>Some literature starts at this time to change the labels of the model parameters, for example the factor effect for level <span class="math inline">\(i\)</span> is defined as <span class="math inline">\(\alpha_i\)</span> instead. <span class="math display">\[
  \begin{aligned}
    Y_{ij} = \mu + \sum_{i = 1}^{A-1}\alpha_i + \varepsilon_{ij}
  \end{aligned}
\]</span></p>
<p>Throughout this book we will keep on using the regression formulation and define each instance of <span class="math inline">\(X_{ij}\)</span> to clearly separate the models from one another.</p>
</div>
</div>
</section>
</section>
<section id="the-anova-model" class="level2" data-number="19.2">
<h2 data-number="19.2" class="anchored" data-anchor-id="the-anova-model"><span class="header-section-number">19.2</span> The ANOVA-model</h2>
<p>The tools of <a href="25_Statistical_inference_two.html" class="quarto-xref"><span>Chapter 14</span></a> allow us to compare two groups of data. But what do we do when we have more than two groups? As an example, we might wish to know how different treatments influence plant growth, as measured by dry weight. If there are two treatments, then there will be three (instead of two) groups, because the treatments will be compared with a control group which does not receive treatment.</p>
<p>Such a dataset is in fact built into R, and is called <code>PlantGrowth</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>PlantGrowth <span class="sc">|&gt;</span> </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span> </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="at">n =</span> <span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 2
   weight group
    &lt;dbl&gt; &lt;fct&gt;
 1   4.17 ctrl 
 2   5.58 ctrl 
 3   5.18 ctrl 
 4   6.11 ctrl 
 5   4.5  ctrl 
 6   4.61 ctrl 
 7   5.17 ctrl 
 8   4.53 ctrl 
 9   5.33 ctrl 
10   5.14 ctrl 
11   4.81 trt1 
12   4.17 trt1 
13   4.41 trt1 
14   3.59 trt1 
15   5.87 trt1 
16   3.83 trt1 
17   6.03 trt1 
18   4.89 trt1 
19   4.32 trt1 
20   4.69 trt1 
21   6.31 trt2 
22   5.12 trt2 
23   5.54 trt2 
24   5.5  trt2 
25   5.37 trt2 
26   5.29 trt2 
27   4.92 trt2 
28   6.15 trt2 
29   5.8  trt2 
30   5.26 trt2 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>PlantGrowth</code> data are in the form of a data frame instead of a tibble; hence we convert it above by using <code>as_tibble</code>.</p>
</div>
</div>
<p>We see that each group consists of ten observations, and we have three groups: the control (<code>ctrl</code>) treatment 1 (<code>trt1</code>), and treatment 2 (<code>trt2</code>). As usual, we visualize the data first, before doing any modeling or tests:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>PlantGrowth <span class="sc">|&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">color =</span> <span class="st">"steelblue"</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">width =</span> <span class="fl">0.05</span>, <span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="43_Anova_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There are two questions one can ask, and they naturally build on each other.</p>
<p>First, we might wonder if the observed differences between the distribution of the data in the different groups is meaningful. Could it be that all data points are actually drawn from the same underlying distribution, and the observed differences are simply due to chance? If so, what is the likelihood of that situation?</p>
<p>And second, provided we can reject the idea that the data in the different groups are identically distributed, which groups are the ones that differ from one another? For example, based on the figure above, it appears reasonable to expect a real difference between treatment 1 and treatment 2, but it is unclear whether the control truly differs from any of the treatments. What we will in practice be calculating is the difference in means between all three groups, whether they is any at all, and if so, which ones differ from each other.</p>
<p>Assuming we are using the factor effect coding, the model can be given as:</p>
<p><span id="eq-factor-anova-example"><span class="math display">\[
\begin{aligned}
    Y_{ij} = \mu + \beta_1 \cdot X_{1j} + \beta_{2}\cdot X_{2j} + \varepsilon_{ij}
\end{aligned}
\tag{19.3}\]</span></span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
    \mu &amp;= \frac{\mu_1 + \mu_2 + \mu_3}{3} \\
    X_{1j} &amp;=
    \begin{cases}
        1 \quad &amp;\text{if control}\\
        -1      &amp;\text{if treatment 2}\\
        0       &amp;\text{otherwise}
    \end{cases} \\
    X_{2j} &amp;=
    \begin{cases}
        1 \quad &amp;\text{if treatment 1}\\
        -1      &amp;\text{if treatment 2}\\
        0       &amp;\text{otherwise}
    \end{cases}
\end{aligned}
\]</span> and <span class="math inline">\(\varepsilon_{ij} \sim N(0, \sigma^2)\)</span>.</p>
<p>Before moving on the with model estimation we need to make sure that R reads the variable containing our factor of interest as a <code>factor</code> type. This can be seen in the tibble output where <code>group</code> is seen as a <code>fct</code>, a factor variable. In order to create the factor effect coding we make use of the function <code>contr.sum(A)</code> where A is the number of levels of the factor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Shows the factor effect coding matrix</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">contr.sum</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [,1] [,2]
1    1    0
2    0    1
3   -1   -1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Shows the current/default coding matrix for the factor</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">contrasts</span>(PlantGrowth<span class="sc">$</span>group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     trt1 trt2
ctrl    0    0
trt1    1    0
trt2    0    1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Saves the factor effect coding matrix into the factor</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">contrasts</span>(PlantGrowth<span class="sc">$</span>group) <span class="ot">&lt;-</span> <span class="fu">contr.sum</span>(<span class="dv">3</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Shows the new coding matrix</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">contrasts</span>(PlantGrowth<span class="sc">$</span>group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
ctrl    1    0
trt1    0    1
trt2   -1   -1</code></pre>
</div>
</div>
<p>The default order in <code>contr.sum</code> is that the last level is considered the “missing” one, but we could reorder the factor levels (using <code>factor(levels = new order)</code> and/or <code>relevel()</code>) before adjusting the coding matrix to get any order we like.</p>
<p>Perhaps unsurprisingly, the function in R that estimates an ANOVA model is the same <code>lm</code> that we have used in <a href="41_Linear_regression.html" class="quarto-xref"><span>Chapter 17</span></a> for performing linear regression:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lm</span>(weight <span class="sc">~</span> group, <span class="at">data =</span> PlantGrowth) </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>model <span class="sc">|&gt;</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ group, data = PlantGrowth)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.0710 -0.4180 -0.0060  0.2627  1.3690 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   5.0730     0.1138  44.573   &lt;2e-16 ***
group1       -0.0410     0.1610  -0.255   0.8009    
group2       -0.4120     0.1610  -2.560   0.0164 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.6234 on 27 degrees of freedom
Multiple R-squared:  0.2641,    Adjusted R-squared:  0.2096 
F-statistic: 4.846 on 2 and 27 DF,  p-value: 0.01591</code></pre>
</div>
</div>
<p>The output takes some care to interpret correctly: out of the three coefficients listed, <code>(Intercept)</code> is the overall mean weight of all the plants; <code>group1</code> is the <em>difference</em> of the mean weight in the control from the overall mean; and <code>group2</code> is the difference of the mean weight in treatment 1 from the overall mean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diagnosticPlots</span>(model, <span class="at">alpha =</span> <span class="dv">1</span>, <span class="at">bins =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="43_Anova_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="480"></p>
<figcaption>Residual analysis of the PlantGrowth model</figcaption>
</figure>
</div>
</div>
</div>
<p>Given the fact that there are only 30 observations, the interpretation of these plots is going to be approximate. Unless there is clear indications of violations we usually assume that the assumptions are fulfilled. In this specific case, there are no clear indications that we are violating the normality assumption however there is an indication that the variance of the residuals are reduced for higher values of the weight.</p>
<p>At this point in the book we do not have any tools at our disposal to handle this so we will (unfortunately) disregard this observation and assume the residuals are homoscedastic (constant variance).</p>
<!-- !TODO ADD A CHAPTER SOMEWHERE ABOUT BOX-COX TRANSFORMATIONS -->
</section>
<section id="statistical-inference-in-an-anova-model" class="level2" data-number="19.3">
<h2 data-number="19.3" class="anchored" data-anchor-id="statistical-inference-in-an-anova-model"><span class="header-section-number">19.3</span> Statistical inference in an ANOVA model</h2>
<p>Recall back to the two questions we asked ourselves in the exploratory phase; are the observed differences between the groups meaningful, and which groups differ from each other. Assuming that the residual analysis enforces that the model assumptions hold, these two questions can be answered by statistical inference.</p>
<p>To get more information, one can pass the result of <code>lm</code> to a function called <code>anova</code>. Despite its name, the role of this function is not to actually estimate the ANOVA model (that was done by <code>lm</code>), but to display its results using the sum-of-squares table, which is also known as the ANOVA table. In this table, each factor, as well as their interactions (if present), get one row each:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>model <span class="sc">|&gt;</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anova</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: weight
          Df  Sum Sq Mean Sq F value  Pr(&gt;F)  
group      2  3.7663  1.8832  4.8461 0.01591 *
Residuals 27 10.4921  0.3886                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<section id="overall-f-test" class="level3" data-number="19.3.1">
<h3 data-number="19.3.1" class="anchored" data-anchor-id="overall-f-test"><span class="header-section-number">19.3.1</span> Overall F-test</h3>
<p>An F-test can be used to determine whether a factor as a whole has a significant impact on the response variable, similar to that of the F-test in <a href="41_Linear_regression.html#sec-inference" class="quarto-xref"><span>Section 17.4</span></a>. Using the model <a href="#eq-factor-oneway" class="quarto-xref">Equation&nbsp;<span>19.2</span></a> we can write the hypotheses as:</p>
<p><span class="math display">\[
\begin{aligned}
    H_0&amp;: \beta_1 = \beta_2 = \dots = \beta_{A-1} = 0 \\
    H_A&amp;: \text{At least one of } \beta_i \text{ in } H_0 \ne 0
\end{aligned}
\]</span></p>
<p>The null hypothesis can be interpreted as no factor level mean differ from the overall mean if all model parameters are 0. In practice this means that the factor has no effect on the response variable. The alternative hypothesis indicate that at least one factor level mean differ from the overall mean, which in contrast results in the factor having an effect on the response variable.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since we can simplify the model to a mean for each factor level, the hypotheses are equivalent to: <span class="math display">\[
\begin{aligned}
    H_0&amp;: \mu_1 = \mu_2 = \dots = \mu_{A-1} = \mu_A \\
    H_A&amp;: \text{At least two of } \mu_i \text{ in } H_0 \text{ differ}
\end{aligned}
\]</span> This can be seen as an extension of the two sample t-test presented in <a href="25_Statistical_inference_two.html#sec-example-comparison" class="quarto-xref"><span>Section 14.1</span></a>.</p>
</div>
</div>
<p>This is where we are starting to see the reason for the name “Analysis of Variance”. Even though we are comparing means between groups, the complexity of actually comparing the means as we would in a two sample t-test would be all to complicated. We can instead make use of the sources of variation introduced in <a href="41_Linear_regression.html#sec-regression-anova" class="quarto-xref"><span>Section 17.4.1</span></a> that simplified for an ANOVA-model summarize the differences in means.</p>
<p><span id="eq-anova-sum-squares"><span class="math display">\[
\begin{aligned}
  SSY &amp;= \sum_{i=1}^A\sum_{j=1}^{n_i}{(Y_{ij} - \overline{Y})^2} \\
  SSA &amp;= \sum_{i=1}^A{n_i \cdot (\overline{Y}_i - \overline{Y})^2} \\
  SSE &amp;= \sum_{i=1}^A\sum_{j=1}^{n_i}{(Y_{ij} - \overline{Y}_i)^2}
\end{aligned}
\tag{19.4}\]</span></span> where</p>
<ul>
<li><span class="math inline">\(SSY\)</span> is the total variation of the response variable written as the difference between each observation <span class="math inline">\(Y_{ij}\)</span> and the total mean <span class="math inline">\(\overline{Y}\)</span>.</li>
<li><span class="math inline">\(SSA\)</span> is the explained variation of each of the factor level means written as the difference between the level mean <span class="math inline">\(\overline{Y}_{i}\)</span> and the total mean <span class="math inline">\(\overline{Y}\)</span>.</li>
<li><span class="math inline">\(SSE\)</span> is the unexplained variation written as the difference between each observation <span class="math inline">\(Y_{ij}\)</span> and the level mean <span class="math inline">\(\overline{Y}_{i}\)</span>.</li>
</ul>
<p>If <span class="math inline">\(H_0\)</span> was true the difference in factor level mean and overall mean would be 0, so <span class="math inline">\(SSA\)</span> would be 0. On the other hand if <span class="math inline">\(H_A\)</span> was true we would see a non-zero sum in <span class="math inline">\(SSA\)</span>. In order to assess whether the difference in means is large enough to not attribute it to chance, we define the test variable as:</p>
<p><span class="math display">\[
  \begin{aligned}
    F_{test} = \frac{SSA / df_A}{SSE / df_E} = \frac{MSA}{MSE}
  \end{aligned}
\]</span> where <span class="math inline">\(df_A\)</span> and <span class="math inline">\(df_E\)</span> are the degrees of freedom of the factor and error respectively. In the ANOVA table output we can see the result of this calculation in the <code>F value</code> column and its corresponding p-value in the <code>Pr(&gt;F)</code> column.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Below is an interactive visualization using simulations of the relationship between the two sources of variation. The simulation assumes we have four groups with known population means<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> we draw a sample from, visualize each group’s distribution and mean (black line) against the overall mean (red line), and finally calculate the sum of squares, the test variable and the p-value.</p>
<pre class="shinylive-r" data-engine="r"><code>#| '!! shinylive warning !!': |
#|   shinylive does not work in self-contained HTML documents.
#|   Please set `embed-resources: false` in your metadata.
#| standalone: true
#| viewerHeight: 1000

require(shiny)
require(bslib)
require(ggplot2)
require(tidyr)
require(dplyr)
require(stringr)
require(munsell)

# Define UI for app that draws a histogram ----
ui &lt;- page_sidebar(
  sidebar = sidebar(open = "open",
    sliderInput("group_1", "Population mean Group 1", min = -1, max = 1, value = 0, step = 0.1),
    sliderInput("group_2", "Population mean Group 2", min = -1, max = 1, value = 0, step = 0.1),
    sliderInput("group_3", "Population mean Group 3", min = -1, max = 1, value = 0, step = 0.1),
    sliderInput("group_4", "Population mean Group 4", min = -1, max = 1, value = 0, step = 0.1),
    numericInput("sample_size", "Sample size", value = 30, min = 1, step = 1),
    actionButton("new_sample", "Draw a new sample"),
    uiOutput("variances"),
  ),
  plotOutput("plot", height = "600px", width = "60%")
)

server &lt;- function(input, output, session) {
  data_input &lt;- reactive({
    means &lt;- c(input$group_1, input$group_2, input$group_3, input$group_4)
    seed &lt;- input$new_sample
    set.seed(seed = seed)
    samples &lt;- sapply(means, FUN = function(x) {
        rnorm(n = input$sample_size, mean = x, sd = 1)
    }) %&gt;% 
      as.data.frame()
    data_samples &lt;- pivot_longer(samples, cols = everything()) %&gt;% 
      mutate(
        name = str_replace(name, "V", "Group ")
      )
  })
  
  output$plot &lt;- renderPlot({
    data_samples &lt;- data_input()
    sample_means &lt;- aggregate(value ~ name, data_samples, mean)

    ggplot2::ggplot(data = data_samples) + aes(x = value) + 
    geom_histogram(binwidth = 0.25, color = "black", fill = "steelblue") +
    geom_vline(data = sample_means, aes(xintercept = value), linetype = 2, linewidth = 1.2) +
    geom_vline(aes(xintercept = mean(value)), color = "#d9230f", linewidth = 1) +
            scale_x_continuous(breaks = seq(-20, 20, by = 1)) +
            facet_grid(rows = vars(name)) + theme_bw() +
            theme(strip.text.y = element_text(angle = 0, color = "white", size = 14),
                  strip.background.y = element_rect(fill = "black"),
                  axis.title.y = element_blank()) +
            labs(x = "Y") 
  })

  output$variances &lt;- renderUI({
    data_samples &lt;- data_input()
    sample_means &lt;- aggregate(value ~ name, data_samples, mean)
    sample_sizes &lt;- aggregate(value ~ name, data_samples, length)

    SSY &lt;- sum((data_samples$value - mean(data_samples$value))^2)

    SSR &lt;- sum(sample_sizes$value * (sample_means$value - mean(data_samples$value))^2)
    MSR &lt;- SSR/(nrow(sample_means) - 1)

    SSE &lt;- SSY-SSR
    MSE &lt;- SSE/(nrow(data_samples) - nrow(sample_means))
    
    pvalue &lt;- pf(
      round(MSR / MSE, 3), 
      df1 = (nrow(sample_means) - 1), 
      df2 = (nrow(data_samples) - nrow(sample_means)),
      lower.tail = FALSE
    )

    withMathJax(
      paste("$$SSY = ", round(SSY, 3), "\\\\",
            "SSA = ", round(SSR, 3), "\\\\",
            "SSE = ", round(SSE, 3), "\\\\",
            "F_{test} = ", round(MSR / MSE, 3), "\\\\",
            "\\text{p-value} = ", pvalue %&gt;% round(3), "$$")
    )
  })
}

# Create Shiny app ----
shinyApp(ui = ui, server = server)
</code></pre>
<p>Initially the population means of the four groups are set at the same level and the sample drawn show factor level means very close to the overall mean and each other. If we adjust one (or more) of the population means to something else, we directly see the impact on the drawn sample. The factor level mean is now further away than the overall mean as well as the other level means. By this logic we can see that a change in at least one mean also changes the relation between the explained and unexplained variation.</p>
</div>
</div>
<p>In our model, the p-value of the overall F-test is <span class="math inline">\(0.01591\)</span> which is lower than a significance level of 5%. This means that we can conclude that the observed differences are significant and we expect at least two of the three population means to differ from one another.</p>
</section>
<section id="multiple-comparisons" class="level3" data-number="19.3.2">
<h3 data-number="19.3.2" class="anchored" data-anchor-id="multiple-comparisons"><span class="header-section-number">19.3.2</span> Multiple comparisons</h3>
<p>If <span class="math inline">\(H_0\)</span> is rejected, we do not yet know which factor levels differ from each other just that at least two does. One could look at this problem as doing multiple pairwise t-tests between each pair of factor levels but this causes two main issues if we want to draw conclusions from all tests at once.</p>
<p>First the method of using repeated t-tests works fine in our example, there are some problems with this approach. One is that it can quickly get out of hand, because having <span class="math inline">\(n\)</span> groups means there will be <span class="math inline">\(n (n - 1) / 2\)</span> pairs to consider. For instance, if the number of groups is 12 (not a particularly large number), then there are 66 unique pairs already. It would not be pleasant to perform this many tests, even if any single test is quite simple to run.</p>
<p>The second relates to the risk of making the wrong decision. Recall the Type I and II errors discussed in <a href="23_Statistical_inference.html#sec-inference-errors" class="quarto-xref"><span>Section 12.2.1.3</span></a>. Using a 5% level of significance, we have a 5% chance of our sample being one of the infinite amount of samples that rejects this specific <span class="math inline">\(H_0\)</span> even though it is true. In practice this means that we risk having observed differences that occur purely by chance. Very simply put, the problem is that if sufficiently many groups are compared, then we might find at least one pair with a low p-value—not because the null hypothesis is false, but because across a large number of observations some p-values will turn out lower than others just by chance. The p-values measure, in effect, the probability that the observed differences are too stark to be due to simple coincidence. But if we create sufficiently many opportunities for such a coincidence to arise, then of course one eventually will. One of the best explanations of this point is in the following cartoon by <a href="https://xkcd.com/">xkcd</a>:</p>
<p><img src="significant.png" class="img-fluid"></p>
<p>Fortunately, there is a way of solving both the problem of automating many pairwise comparisons, as well as adjusting the p-values to account for multiple testing. The way forward is to perform a <em>post hoc</em> (Latin “after this”) test.</p>
<section id="tukey-family-confidence" class="level4" data-number="19.3.2.1">
<h4 data-number="19.3.2.1" class="anchored" data-anchor-id="tukey-family-confidence"><span class="header-section-number">19.3.2.1</span> Tukey family confidence</h4>
<p>The Tukey(-Kramer) method is used to adjust the p-values of all pairwise comparisions of factor level means. We will make use of the package <code>emmeans</code> and its different functions to first create a table of the estimated means and their corresponding standard errors and then the pairwise comparisons with the Tukey post-hoc adjustment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(emmeans)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Calculates the factor level means of the specified factor</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>factorMeans <span class="ot">&lt;-</span> <span class="fu">emmeans</span>(model, <span class="at">specs =</span> <span class="sc">~</span> group)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Calculates the pairwise comparisons with a Tukey adjustment</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>tukey <span class="ot">&lt;-</span> <span class="fu">pairs</span>(factorMeans, <span class="at">adjust =</span> <span class="st">"tukey"</span>) </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Presents the pairwise comparisons sorted by the estimated difference</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>tukey <span class="sc">|&gt;</span> </span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()  <span class="sc">|&gt;</span>  </span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">abs</span>(estimate) <span class="sc">|&gt;</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">desc</span>()</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  contrast    estimate    SE    df t.ratio p.value
  &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 trt1 - trt2   -0.865 0.279    27   -3.10  0.0120
2 ctrl - trt2   -0.494 0.279    27   -1.77  0.198 
3 ctrl - trt1    0.371 0.279    27    1.33  0.391 </code></pre>
</div>
</div>
<p>The table shows three <em>contrasts</em>, which is the three pairwise comparisons between each factor level and the adjusted p-values of each comparison. The hypotheses tested are:</p>
<p><span class="math display">\[
  \begin{aligned}
  H_0&amp;: \mu_i - \mu_{j} = 0\\
  H_A&amp;: \mu_i - \mu_{j} \ne 0\\
  \end{aligned}
\]</span> where <span class="math inline">\(i \ne j\)</span>.</p>
<p>What has the post hoc Tukey test revealed? Precisely what we have been suspecting: that the only difference worth noting is the one between the two treatments (first row of the table, where the adjusted p-value is sufficiently small to have a chance of pointing at a real difference).</p>
</section>
<section id="dunnetts-test-against-control" class="level4" data-number="19.3.2.2">
<h4 data-number="19.3.2.2" class="anchored" data-anchor-id="dunnetts-test-against-control"><span class="header-section-number">19.3.2.2</span> Dunnett’s test against control</h4>
<p>Considering we have a control group, there exist a specific test that looks at all pairwise comparisons of factor levels only against the control group. This is called a <em>Dunnett’s</em> test.</p>
<p>Changing the method argument to <code>dunnett</code> and specifying which factor level contains the control group in <code>ref</code> we now focus on the two comparisons against the control.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">contrast</span>(factorMeans, <span class="at">method =</span> <span class="st">"dunnett"</span>, <span class="at">ref =</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">abs</span>(estimate) <span class="sc">|&gt;</span> </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">desc</span>()</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 6
  contrast    estimate    SE    df t.ratio p.value
  &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 trt2 - ctrl    0.494 0.279    27    1.77   0.158
2 trt1 - ctrl   -0.371 0.279    27   -1.33   0.330</code></pre>
</div>
</div>
<p>It shows that neither treatment has significant differences against the control on a 5% significance level.</p>
</section>
</section>
</section>
<section id="the-kruskalwallis-and-dunn-tests" class="level2" data-number="19.4">
<h2 data-number="19.4" class="anchored" data-anchor-id="the-kruskalwallis-and-dunn-tests"><span class="header-section-number">19.4</span> The Kruskal–Wallis and Dunn tests</h2>
<p>What happens if the regression model assumptions are not fulfilled? Starting with the first of these questions, there exists an analogue to the Wilcoxon rank sum test which works when there are more than two groups of data. This is the <em>Kruskal–Wallis test</em><a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> , which can be used with any number of groups as long as those groups vary within a single factor.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></p>
<p><em>The Kruskal–Wallis</em> test is non-parametric, and therefore does not rely on assumptions such as the normality of the residuals. Its implementation in R, <code>kruskal.test</code>, is analogous to <code>wilcox.test</code>, <code>t.test</code> <code>lm</code>, or <code>mblm</code>: it takes a formula and the data as inputs. Therefore, to perform the test on the <code>PlantGrowth</code> data, we write:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">kruskal.test</span>(weight <span class="sc">~</span> group, <span class="at">data =</span> PlantGrowth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Kruskal-Wallis rank sum test

data:  weight by group
Kruskal-Wallis chi-squared = 7.9882, df = 2, p-value = 0.01842</code></pre>
</div>
</div>
<p>The null hypothesis of the Kruskal–Wallis test is that the observations from all groups were sampled from the same underlying distribution—that is, that there are no differences between the groups other than those attributed to random noise. Consequently, when the p-value is low (like above), this means that it is unlikely that the data in all groups come from the same distribution, and thus that <em>at least</em> one group differs from the others.</p>
<p>The p-value above, in the row belonging to the factor <code>group</code>, under the column <code>Pr(&gt;F)</code>, is the analogue of the p-value calculated by the F-test (which was 0.01591). We can see that the two tests agree qualitatively.</p>
<p>In this case, the <em>Dunn test</em> is the non-parametric post hoc test we want to use to determine which group(s) differ. This test is implemented in R, but not in any of the basic packages. To use it, one must first install the <code>FSA</code> package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"FSA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once it is installed, the package should be loaded:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(FSA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And then, the Dunn test (<code>dunnTest</code>) follows the familiar syntax of receiving a formula and the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dunnTest</span>(weight <span class="sc">~</span> group, <span class="at">data =</span> PlantGrowth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Dunn (1964) Kruskal-Wallis multiple comparison</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>  p-values adjusted with the Holm method.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Comparison         Z    P.unadj      P.adj
1 ctrl - trt1  1.117725 0.26368427 0.26368427
2 ctrl - trt2 -1.689290 0.09116394 0.18232789
3 trt1 - trt2 -2.807015 0.00500029 0.01500087</code></pre>
</div>
</div>
<p>The table above is the output of the test, and has four columns. The first column shows which two groups are being compared. The next column, called <code>Z</code>, is the value of the test statistic, which we need not concern ourselves with here. Next, we have the unadjusted p-values; and finally, the adjusted p-values (<code>P.adj</code>), which have been corrected to account for the multiple testing problem mentioned above. Therefore, the adjusted p-values will always be as large or larger than the unadjusted ones.</p>
<p>What has the post hoc Dunn test revealed? Precisely what we have been suspecting: that the only difference worth noting is the one between the two treatments (last row of the table, where the adjusted p-value is sufficiently small to have a chance of pointing at a real difference).</p>
<p>In this case, the results from the Dunn test and the Tukey test are in agreement: only the difference between the two treatment groups stands out as having a reasonable chance of being real.</p>
</section>
<section id="exercises" class="level2" data-number="19.5">
<h2 data-number="19.5" class="anchored" data-anchor-id="exercises"><span class="header-section-number">19.5</span> Exercises</h2>
<ol type="1">
<li>The file <a href="https://raw.githubusercontent.com/dysordys/intro-R-and-stats/main/data/daphnia_growth.zip"><code>daphnia_growth.csv</code></a> contains data on the growth rate of <em>Daphnia</em> populations that are infected with various parasites. There are four groups of observations: the control (no parasites), infection with <em>Metschnikowia bicuspidata</em>, infection with <em>Pansporella perplexa</em>, and finally, infection with <em>Pasteuria ramosa</em>. Each group has ten replicate observations. Are growth rates affected by parasite load?
<ul>
<li>Before doing any tests, visualize and explore the data, and make sure you have a solid expectation for the results of any statistical analysis.</li>
<li>Answer the question whether growth rates affected by parasite load by first applying a non-parametric test (and a subsequent non-parametric post-hoc test if needed).</li>
<li>Next, apply a parametric test in the same way: by applying the test and running post-hoc tests if needed.</li>
<li>Do not forget to create diagnostic plots, to see if the assumptions behind the parametric test are satisfied to an acceptable degree. Is that the case? And do the results from the parametric and non-parametric tests agree with one another?</li>
</ul></li>
<li>In <a href="https://raw.githubusercontent.com/dysordys/intro-R-and-stats/main/data/ponds.zip"><code>ponds.csv</code></a>, measured acidity data (pH) is reported from four different ponds. Do the ponds differ in acidity, and if so, which ones from which others? Answer using both non-parametric and parametric tests, with appropriate post-hoc analyses. Check whether these different methods of analysis agree, and make sure that the assumptions behind the parametric test are satisfied using diagnostic plots. (Note: in this dataset, some values are missing.)</li>
</ol>


</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p>Each group is actually <span class="math inline">\(N(\mu_i, \sigma^2 = 1)\)</span>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The Kruskal–Wallis test is sometimes referred to as “non-parametric ANOVA”. While this is perfectly fine as a label, one should be aware that it is, strictly speaking, a misnomer: the Kruskal–Wallis test does not rely on computing variances at all.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>In <a href="44_Anova_two_way.html" class="quarto-xref"><span>Chapter 20</span></a> we will see examples where multiple independent factors are varied, and each possible combination results in a separate group. For example, if the effects of three different dosages of vitamin C are examined on the tooth growth of Guinea pigs, and the vitamin is also supplied in two distinct forms of either orange juice or raw ascorbic acid, then there will be <span class="math inline">\(3 \cdot 2 = 6\)</span> groups, defined by the two factors of dosage and form of supplement.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./42_Multiple_and_Nonlinear_regression.html" class="pagination-link" aria-label="Advanced regression topics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Advanced regression topics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./44_Anova_two_way.html" class="pagination-link" aria-label="Two-way ANOVA and the Scheirer--Ray--Hare test">
        <span class="nav-page-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Two-way ANOVA and the Scheirer–Ray–Hare test</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>